// @patternfly/react-styles/css/components/DataList/data-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./data-list.css");
exports.default = {
    dataList: 'pf-c-data-list',
    dataListItemAction: 'pf-c-data-list__item-action',
    dataListItem: 'pf-c-data-list__item',
    dataListItemRow: 'pf-c-data-list__item-row',
    dataListItemControl: 'pf-c-data-list__item-control',
    dataListAction: 'pf-c-data-list__action',
    dataListToggle: 'pf-c-data-list__toggle',
    button: 'pf-c-button',
    dataListExpandableContent: 'pf-c-data-list__expandable-content',
    dataListItemContent: 'pf-c-data-list__item-content',
    dataListCell: 'pf-c-data-list__cell',
    dataListExpandableContentBody: 'pf-c-data-list__expandable-content-body',
    modifiers: {
        'hidden': 'pf-m-hidden',
        'hiddenOnSm': 'pf-m-hidden-on-sm',
        'visibleOnSm': 'pf-m-visible-on-sm',
        'hiddenOnMd': 'pf-m-hidden-on-md',
        'visibleOnMd': 'pf-m-visible-on-md',
        'hiddenOnLg': 'pf-m-hidden-on-lg',
        'visibleOnLg': 'pf-m-visible-on-lg',
        'hiddenOnXl': 'pf-m-hidden-on-xl',
        'visibleOnXl': 'pf-m-visible-on-xl',
        'hiddenOn_2xl': 'pf-m-hidden-on-2xl',
        'visibleOn_2xl': 'pf-m-visible-on-2xl',
        'compact': 'pf-m-compact',
        'expanded': 'pf-m-expanded',
        'selected': 'pf-m-selected',
        'selectable': 'pf-m-selectable',
        'icon': 'pf-m-icon',
        'noFill': 'pf-m-no-fill',
        'alignRight': 'pf-m-align-right',
        'flex_2': 'pf-m-flex-2',
        'flex_3': 'pf-m-flex-3',
        'flex_4': 'pf-m-flex-4',
        'flex_5': 'pf-m-flex-5',
        'noPadding': 'pf-m-no-padding'
    }
};
