// @patternfly/react-styles/css/components/AppLauncher/app-launcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./app-launcher.css");
exports.default = {
    appLauncher: 'pf-c-app-launcher',
    appLauncherToggle: 'pf-c-app-launcher__toggle',
    divider: 'pf-c-divider',
    appLauncherMenu: 'pf-c-app-launcher__menu',
    appLauncherMenuSearch: 'pf-c-app-launcher__menu-search',
    appLauncherMenuWrapper: 'pf-c-app-launcher__menu-wrapper',
    appLauncherMenuItem: 'pf-c-app-launcher__menu-item',
    appLauncherMenuItemExternalIcon: 'pf-c-app-launcher__menu-item-external-icon',
    appLauncherMenuItemIcon: 'pf-c-app-launcher__menu-item-icon',
    appLauncherSeparator: 'pf-c-app-launcher__separator',
    appLauncherGroup: 'pf-c-app-launcher__group',
    appLauncherGroupTitle: 'pf-c-app-launcher__group-title',
    modifiers: {
        'expanded': 'pf-m-expanded',
        'hover': 'pf-m-hover',
        'active': 'pf-m-active',
        'focus': 'pf-m-focus',
        'disabled': 'pf-m-disabled',
        'alignRight': 'pf-m-align-right',
        'top': 'pf-m-top',
        'favorite': 'pf-m-favorite',
        'external': 'pf-m-external',
        'link': 'pf-m-link',
        'action': 'pf-m-action'
    }
};
