// @patternfly/react-styles/css/components/Login/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./login.css");
exports.default = {
    loginHeader: 'pf-c-login__header',
    loginFooter: 'pf-c-login__footer',
    card: 'pf-c-card',
    button: 'pf-c-button',
    login: 'pf-c-login',
    loginContainer: 'pf-c-login__container',
    brand: 'pf-c-brand',
    loginMain: 'pf-c-login__main',
    loginMainHeader: 'pf-c-login__main-header',
    loginMainFooter: 'pf-c-login__main-footer',
    dropdown: 'pf-c-dropdown',
    loginMainHeaderDesc: 'pf-c-login__main-header-desc',
    loginMainBody: 'pf-c-login__main-body',
    formHelperText: 'pf-c-form__helper-text',
    formHelperTextIcon: 'pf-c-form__helper-text-icon',
    title: 'pf-c-title',
    loginMainFooterLinks: 'pf-c-login__main-footer-links',
    loginMainFooterLinksItem: 'pf-c-login__main-footer-links-item',
    loginMainFooterLinksItemLink: 'pf-c-login__main-footer-links-item-link',
    loginMainFooterBand: 'pf-c-login__main-footer-band',
    list: 'pf-c-list',
    modifiers: {}
};
