// @patternfly/react-styles/css/utilities/Sizing/sizing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./sizing.css");
exports.default = {
    wAuto: 'pf-u-w-auto',
    wInitial: 'pf-u-w-initial',
    wInherit: 'pf-u-w-inherit',
    w_0: 'pf-u-w-0',
    w_25: 'pf-u-w-25',
    w_33: 'pf-u-w-33',
    w_50: 'pf-u-w-50',
    w_66: 'pf-u-w-66',
    w_75: 'pf-u-w-75',
    w_100: 'pf-u-w-100',
    w_25vw: 'pf-u-w-25vw',
    w_33vw: 'pf-u-w-33vw',
    w_50vw: 'pf-u-w-50vw',
    w_66vw: 'pf-u-w-66vw',
    w_75vw: 'pf-u-w-75vw',
    w_100vw: 'pf-u-w-100vw',
    wAutoOnSm: 'pf-u-w-auto-on-sm',
    wInitialOnSm: 'pf-u-w-initial-on-sm',
    wInheritOnSm: 'pf-u-w-inherit-on-sm',
    w_0OnSm: 'pf-u-w-0-on-sm',
    w_25OnSm: 'pf-u-w-25-on-sm',
    w_33OnSm: 'pf-u-w-33-on-sm',
    w_50OnSm: 'pf-u-w-50-on-sm',
    w_66OnSm: 'pf-u-w-66-on-sm',
    w_75OnSm: 'pf-u-w-75-on-sm',
    w_100OnSm: 'pf-u-w-100-on-sm',
    w_25vwOnSm: 'pf-u-w-25vw-on-sm',
    w_33vwOnSm: 'pf-u-w-33vw-on-sm',
    w_50vwOnSm: 'pf-u-w-50vw-on-sm',
    w_66vwOnSm: 'pf-u-w-66vw-on-sm',
    w_75vwOnSm: 'pf-u-w-75vw-on-sm',
    w_100vwOnSm: 'pf-u-w-100vw-on-sm',
    wAutoOnMd: 'pf-u-w-auto-on-md',
    wInitialOnMd: 'pf-u-w-initial-on-md',
    wInheritOnMd: 'pf-u-w-inherit-on-md',
    w_0OnMd: 'pf-u-w-0-on-md',
    w_25OnMd: 'pf-u-w-25-on-md',
    w_33OnMd: 'pf-u-w-33-on-md',
    w_50OnMd: 'pf-u-w-50-on-md',
    w_66OnMd: 'pf-u-w-66-on-md',
    w_75OnMd: 'pf-u-w-75-on-md',
    w_100OnMd: 'pf-u-w-100-on-md',
    w_25vwOnMd: 'pf-u-w-25vw-on-md',
    w_33vwOnMd: 'pf-u-w-33vw-on-md',
    w_50vwOnMd: 'pf-u-w-50vw-on-md',
    w_66vwOnMd: 'pf-u-w-66vw-on-md',
    w_75vwOnMd: 'pf-u-w-75vw-on-md',
    w_100vwOnMd: 'pf-u-w-100vw-on-md',
    wAutoOnLg: 'pf-u-w-auto-on-lg',
    wInitialOnLg: 'pf-u-w-initial-on-lg',
    wInheritOnLg: 'pf-u-w-inherit-on-lg',
    w_0OnLg: 'pf-u-w-0-on-lg',
    w_25OnLg: 'pf-u-w-25-on-lg',
    w_33OnLg: 'pf-u-w-33-on-lg',
    w_50OnLg: 'pf-u-w-50-on-lg',
    w_66OnLg: 'pf-u-w-66-on-lg',
    w_75OnLg: 'pf-u-w-75-on-lg',
    w_100OnLg: 'pf-u-w-100-on-lg',
    w_25vwOnLg: 'pf-u-w-25vw-on-lg',
    w_33vwOnLg: 'pf-u-w-33vw-on-lg',
    w_50vwOnLg: 'pf-u-w-50vw-on-lg',
    w_66vwOnLg: 'pf-u-w-66vw-on-lg',
    w_75vwOnLg: 'pf-u-w-75vw-on-lg',
    w_100vwOnLg: 'pf-u-w-100vw-on-lg',
    wAutoOnXl: 'pf-u-w-auto-on-xl',
    wInitialOnXl: 'pf-u-w-initial-on-xl',
    wInheritOnXl: 'pf-u-w-inherit-on-xl',
    w_0OnXl: 'pf-u-w-0-on-xl',
    w_25OnXl: 'pf-u-w-25-on-xl',
    w_33OnXl: 'pf-u-w-33-on-xl',
    w_50OnXl: 'pf-u-w-50-on-xl',
    w_66OnXl: 'pf-u-w-66-on-xl',
    w_75OnXl: 'pf-u-w-75-on-xl',
    w_100OnXl: 'pf-u-w-100-on-xl',
    w_25vwOnXl: 'pf-u-w-25vw-on-xl',
    w_33vwOnXl: 'pf-u-w-33vw-on-xl',
    w_50vwOnXl: 'pf-u-w-50vw-on-xl',
    w_66vwOnXl: 'pf-u-w-66vw-on-xl',
    w_75vwOnXl: 'pf-u-w-75vw-on-xl',
    w_100vwOnXl: 'pf-u-w-100vw-on-xl',
    wAutoOn_2xl: 'pf-u-w-auto-on-2xl',
    wInitialOn_2xl: 'pf-u-w-initial-on-2xl',
    wInheritOn_2xl: 'pf-u-w-inherit-on-2xl',
    w_0On_2xl: 'pf-u-w-0-on-2xl',
    w_25On_2xl: 'pf-u-w-25-on-2xl',
    w_33On_2xl: 'pf-u-w-33-on-2xl',
    w_50On_2xl: 'pf-u-w-50-on-2xl',
    w_66On_2xl: 'pf-u-w-66-on-2xl',
    w_75On_2xl: 'pf-u-w-75-on-2xl',
    w_100On_2xl: 'pf-u-w-100-on-2xl',
    w_25vwOn_2xl: 'pf-u-w-25vw-on-2xl',
    w_33vwOn_2xl: 'pf-u-w-33vw-on-2xl',
    w_50vwOn_2xl: 'pf-u-w-50vw-on-2xl',
    w_66vwOn_2xl: 'pf-u-w-66vw-on-2xl',
    w_75vwOn_2xl: 'pf-u-w-75vw-on-2xl',
    w_100vwOn_2xl: 'pf-u-w-100vw-on-2xl',
    hAuto: 'pf-u-h-auto',
    hInitial: 'pf-u-h-initial',
    hInherit: 'pf-u-h-inherit',
    h_0: 'pf-u-h-0',
    h_25: 'pf-u-h-25',
    h_33: 'pf-u-h-33',
    h_50: 'pf-u-h-50',
    h_66: 'pf-u-h-66',
    h_75: 'pf-u-h-75',
    h_100: 'pf-u-h-100',
    h_25vh: 'pf-u-h-25vh',
    h_33vh: 'pf-u-h-33vh',
    h_50vh: 'pf-u-h-50vh',
    h_66vh: 'pf-u-h-66vh',
    h_75vh: 'pf-u-h-75vh',
    h_100vh: 'pf-u-h-100vh',
    hAutoOnSm: 'pf-u-h-auto-on-sm',
    hInitialOnSm: 'pf-u-h-initial-on-sm',
    hInheritOnSm: 'pf-u-h-inherit-on-sm',
    h_0OnSm: 'pf-u-h-0-on-sm',
    h_25OnSm: 'pf-u-h-25-on-sm',
    h_33OnSm: 'pf-u-h-33-on-sm',
    h_50OnSm: 'pf-u-h-50-on-sm',
    h_66OnSm: 'pf-u-h-66-on-sm',
    h_75OnSm: 'pf-u-h-75-on-sm',
    h_100OnSm: 'pf-u-h-100-on-sm',
    h_25vhOnSm: 'pf-u-h-25vh-on-sm',
    h_33vhOnSm: 'pf-u-h-33vh-on-sm',
    h_50vhOnSm: 'pf-u-h-50vh-on-sm',
    h_66vhOnSm: 'pf-u-h-66vh-on-sm',
    h_75vhOnSm: 'pf-u-h-75vh-on-sm',
    h_100vhOnSm: 'pf-u-h-100vh-on-sm',
    hAutoOnMd: 'pf-u-h-auto-on-md',
    hInitialOnMd: 'pf-u-h-initial-on-md',
    hInheritOnMd: 'pf-u-h-inherit-on-md',
    h_0OnMd: 'pf-u-h-0-on-md',
    h_25OnMd: 'pf-u-h-25-on-md',
    h_33OnMd: 'pf-u-h-33-on-md',
    h_50OnMd: 'pf-u-h-50-on-md',
    h_66OnMd: 'pf-u-h-66-on-md',
    h_75OnMd: 'pf-u-h-75-on-md',
    h_100OnMd: 'pf-u-h-100-on-md',
    h_25vhOnMd: 'pf-u-h-25vh-on-md',
    h_33vhOnMd: 'pf-u-h-33vh-on-md',
    h_50vhOnMd: 'pf-u-h-50vh-on-md',
    h_66vhOnMd: 'pf-u-h-66vh-on-md',
    h_75vhOnMd: 'pf-u-h-75vh-on-md',
    h_100vhOnMd: 'pf-u-h-100vh-on-md',
    hAutoOnLg: 'pf-u-h-auto-on-lg',
    hInitialOnLg: 'pf-u-h-initial-on-lg',
    hInheritOnLg: 'pf-u-h-inherit-on-lg',
    h_0OnLg: 'pf-u-h-0-on-lg',
    h_25OnLg: 'pf-u-h-25-on-lg',
    h_33OnLg: 'pf-u-h-33-on-lg',
    h_50OnLg: 'pf-u-h-50-on-lg',
    h_66OnLg: 'pf-u-h-66-on-lg',
    h_75OnLg: 'pf-u-h-75-on-lg',
    h_100OnLg: 'pf-u-h-100-on-lg',
    h_25vhOnLg: 'pf-u-h-25vh-on-lg',
    h_33vhOnLg: 'pf-u-h-33vh-on-lg',
    h_50vhOnLg: 'pf-u-h-50vh-on-lg',
    h_66vhOnLg: 'pf-u-h-66vh-on-lg',
    h_75vhOnLg: 'pf-u-h-75vh-on-lg',
    h_100vhOnLg: 'pf-u-h-100vh-on-lg',
    hAutoOnXl: 'pf-u-h-auto-on-xl',
    hInitialOnXl: 'pf-u-h-initial-on-xl',
    hInheritOnXl: 'pf-u-h-inherit-on-xl',
    h_0OnXl: 'pf-u-h-0-on-xl',
    h_25OnXl: 'pf-u-h-25-on-xl',
    h_33OnXl: 'pf-u-h-33-on-xl',
    h_50OnXl: 'pf-u-h-50-on-xl',
    h_66OnXl: 'pf-u-h-66-on-xl',
    h_75OnXl: 'pf-u-h-75-on-xl',
    h_100OnXl: 'pf-u-h-100-on-xl',
    h_25vhOnXl: 'pf-u-h-25vh-on-xl',
    h_33vhOnXl: 'pf-u-h-33vh-on-xl',
    h_50vhOnXl: 'pf-u-h-50vh-on-xl',
    h_66vhOnXl: 'pf-u-h-66vh-on-xl',
    h_75vhOnXl: 'pf-u-h-75vh-on-xl',
    h_100vhOnXl: 'pf-u-h-100vh-on-xl',
    hAutoOn_2xl: 'pf-u-h-auto-on-2xl',
    hInitialOn_2xl: 'pf-u-h-initial-on-2xl',
    hInheritOn_2xl: 'pf-u-h-inherit-on-2xl',
    h_0On_2xl: 'pf-u-h-0-on-2xl',
    h_25On_2xl: 'pf-u-h-25-on-2xl',
    h_33On_2xl: 'pf-u-h-33-on-2xl',
    h_50On_2xl: 'pf-u-h-50-on-2xl',
    h_66On_2xl: 'pf-u-h-66-on-2xl',
    h_75On_2xl: 'pf-u-h-75-on-2xl',
    h_100On_2xl: 'pf-u-h-100-on-2xl',
    h_25vhOn_2xl: 'pf-u-h-25vh-on-2xl',
    h_33vhOn_2xl: 'pf-u-h-33vh-on-2xl',
    h_50vhOn_2xl: 'pf-u-h-50vh-on-2xl',
    h_66vhOn_2xl: 'pf-u-h-66vh-on-2xl',
    h_75vhOn_2xl: 'pf-u-h-75vh-on-2xl',
    h_100vhOn_2xl: 'pf-u-h-100vh-on-2xl',
    modifiers: {}
};
