// @patternfly/react-styles/css/components/OptionsMenu/options-menu.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./options-menu.css");
exports.default = {
    optionsMenu: 'pf-c-options-menu',
    divider: 'pf-c-divider',
    optionsMenuToggle: 'pf-c-options-menu__toggle',
    optionsMenuToggleButton: 'pf-c-options-menu__toggle-button',
    optionsMenuToggleText: 'pf-c-options-menu__toggle-text',
    optionsMenuToggleIcon: 'pf-c-options-menu__toggle-icon',
    optionsMenuMenu: 'pf-c-options-menu__menu',
    optionsMenuMenuItem: 'pf-c-options-menu__menu-item',
    optionsMenuMenuItemIcon: 'pf-c-options-menu__menu-item-icon',
    optionsMenuSeparator: 'pf-c-options-menu__separator',
    optionsMenuGroup: 'pf-c-options-menu__group',
    optionsMenuGroupTitle: 'pf-c-options-menu__group-title',
    modifiers: {
        'plain': 'pf-m-plain',
        'text': 'pf-m-text',
        'hover': 'pf-m-hover',
        'active': 'pf-m-active',
        'focus': 'pf-m-focus',
        'expanded': 'pf-m-expanded',
        'disabled': 'pf-m-disabled',
        'top': 'pf-m-top',
        'alignRight': 'pf-m-align-right'
    }
};
