// @patternfly/react-styles/css/components/Nav/nav.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./nav.css");
exports.default = {
    nav: 'pf-c-nav',
    navScrollButton: 'pf-c-nav__scroll-button',
    divider: 'pf-c-divider',
    navItem: 'pf-c-nav__item',
    navLink: 'pf-c-nav__link',
    navList: 'pf-c-nav__list',
    pageSidebar: 'pf-c-page__sidebar',
    pageHeaderNav: 'pf-c-page__header-nav',
    navSeparator: 'pf-c-nav__separator',
    navSimpleList: 'pf-c-nav__simple-list',
    navToggle: 'pf-c-nav__toggle',
    navHorizontalList: 'pf-c-nav__horizontal-list',
    navTertiaryList: 'pf-c-nav__tertiary-list',
    navSubnav: 'pf-c-nav__subnav',
    navSection: 'pf-c-nav__section',
    navSectionTitle: 'pf-c-nav__section-title',
    modifiers: {
        'start': 'pf-m-start',
        'end': 'pf-m-end',
        'dark': 'pf-m-dark',
        'current': 'pf-m-current',
        'expanded': 'pf-m-expanded',
        'hover': 'pf-m-hover',
        'active': 'pf-m-active',
        'focus': 'pf-m-focus'
    }
};
