// @patternfly/react-styles/css/utilities/Alignment/alignment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./alignment.css");
exports.default = {
    textAlignLeft: 'pf-u-text-align-left',
    textAlignCenter: 'pf-u-text-align-center',
    textAlignRight: 'pf-u-text-align-right',
    textAlignJustify: 'pf-u-text-align-justify',
    textAlignLeftOnSm: 'pf-u-text-align-left-on-sm',
    textAlignCenterOnSm: 'pf-u-text-align-center-on-sm',
    textAlignRightOnSm: 'pf-u-text-align-right-on-sm',
    textAlignJustifyOnSm: 'pf-u-text-align-justify-on-sm',
    textAlignLeftOnMd: 'pf-u-text-align-left-on-md',
    textAlignCenterOnMd: 'pf-u-text-align-center-on-md',
    textAlignRightOnMd: 'pf-u-text-align-right-on-md',
    textAlignJustifyOnMd: 'pf-u-text-align-justify-on-md',
    textAlignLeftOnLg: 'pf-u-text-align-left-on-lg',
    textAlignCenterOnLg: 'pf-u-text-align-center-on-lg',
    textAlignRightOnLg: 'pf-u-text-align-right-on-lg',
    textAlignJustifyOnLg: 'pf-u-text-align-justify-on-lg',
    textAlignLeftOnXl: 'pf-u-text-align-left-on-xl',
    textAlignCenterOnXl: 'pf-u-text-align-center-on-xl',
    textAlignRightOnXl: 'pf-u-text-align-right-on-xl',
    textAlignJustifyOnXl: 'pf-u-text-align-justify-on-xl',
    textAlignLeftOn_2xl: 'pf-u-text-align-left-on-2xl',
    textAlignCenterOn_2xl: 'pf-u-text-align-center-on-2xl',
    textAlignRightOn_2xl: 'pf-u-text-align-right-on-2xl',
    textAlignJustifyOn_2xl: 'pf-u-text-align-justify-on-2xl',
    modifiers: {}
};
