// @patternfly/react-styles/css/components/Page/page.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./page.css");
exports.default = {
    pageHeader: 'pf-c-page__header',
    pageSidebar: 'pf-c-page__sidebar',
    pageMainSection: 'pf-c-page__main-section',
    card: 'pf-c-card',
    button: 'pf-c-button',
    page: 'pf-c-page',
    pageHeaderBrand: 'pf-c-page__header-brand',
    pageHeaderBrandLink: 'pf-c-page__header-brand-link',
    brand: 'pf-c-brand',
    pageHeaderBrandToggle: 'pf-c-page__header-brand-toggle',
    pageHeaderNav: 'pf-c-page__header-nav',
    nav: 'pf-c-nav',
    navScrollButton: 'pf-c-nav__scroll-button',
    pageHeaderTools: 'pf-c-page__header-tools',
    notificationBadge: 'pf-c-notification-badge',
    avatar: 'pf-c-avatar',
    pageHeaderToolsGroup: 'pf-c-page__header-tools-group',
    pageSidebarBody: 'pf-c-page__sidebar-body',
    pageMain: 'pf-c-page__main',
    pageDrawer: 'pf-c-page__drawer',
    pageMainNav: 'pf-c-page__main-nav',
    pageMainBreadcrumb: 'pf-c-page__main-breadcrumb',
    pageMainWizard: 'pf-c-page__main-wizard',
    drawer: 'pf-c-drawer',
    modifiers: {
        'dark': 'pf-m-dark',
        'icons': 'pf-m-icons',
        'selected': 'pf-m-selected',
        'unread': 'pf-m-unread',
        'mobile': 'pf-m-mobile',
        'user': 'pf-m-user',
        'expanded': 'pf-m-expanded',
        'collapsed': 'pf-m-collapsed',
        'fill': 'pf-m-fill',
        'noFill': 'pf-m-no-fill',
        'light': 'pf-m-light',
        'dark_100': 'pf-m-dark-100',
        'dark_200': 'pf-m-dark-200',
        'noPadding': 'pf-m-no-padding',
        'noPaddingMobile': 'pf-m-no-padding-mobile'
    }
};
