// @patternfly/react-styles/css/components/Wizard/wizard.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./wizard.css");
exports.default = {
    wizardHeader: 'pf-c-wizard__header',
    card: 'pf-c-card',
    button: 'pf-c-button',
    wizard: 'pf-c-wizard',
    wizardOuterWrap: 'pf-c-wizard__outer-wrap',
    wizardNav: 'pf-c-wizard__nav',
    wizardFooter: 'pf-c-wizard__footer',
    wizardToggle: 'pf-c-wizard__toggle',
    wizardClose: 'pf-c-wizard__close',
    wizardTitle: 'pf-c-wizard__title',
    wizardDescription: 'pf-c-wizard__description',
    wizardToggleIcon: 'pf-c-wizard__toggle-icon',
    wizardToggleList: 'pf-c-wizard__toggle-list',
    wizardToggleListItem: 'pf-c-wizard__toggle-list-item',
    wizardToggleNum: 'pf-c-wizard__toggle-num',
    wizardToggleSeparator: 'pf-c-wizard__toggle-separator',
    wizardInnerWrap: 'pf-c-wizard__inner-wrap',
    wizardNavList: 'pf-c-wizard__nav-list',
    wizardNavLink: 'pf-c-wizard__nav-link',
    wizardNavItem: 'pf-c-wizard__nav-item',
    wizardMain: 'pf-c-wizard__main',
    wizardMainBody: 'pf-c-wizard__main-body',
    modifiers: {
        'fullWidth': 'pf-m-full-width',
        'fullHeight': 'pf-m-full-height',
        'compactNav': 'pf-m-compact-nav',
        'finished': 'pf-m-finished',
        'inPage': 'pf-m-in-page',
        'expanded': 'pf-m-expanded',
        'current': 'pf-m-current',
        'hover': 'pf-m-hover',
        'focus': 'pf-m-focus',
        'disabled': 'pf-m-disabled',
        'noPadding': 'pf-m-no-padding'
    }
};
