// @patternfly/react-styles/css/utilities/Spacing/spacing.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./spacing.css");
exports.default = {
    mAuto: 'pf-u-m-auto',
    m_0: 'pf-u-m-0',
    mXs: 'pf-u-m-xs',
    mSm: 'pf-u-m-sm',
    mMd: 'pf-u-m-md',
    mLg: 'pf-u-m-lg',
    mXl: 'pf-u-m-xl',
    m_2xl: 'pf-u-m-2xl',
    m_3xl: 'pf-u-m-3xl',
    mtAuto: 'pf-u-mt-auto',
    mt_0: 'pf-u-mt-0',
    mtXs: 'pf-u-mt-xs',
    mtSm: 'pf-u-mt-sm',
    mtMd: 'pf-u-mt-md',
    mtLg: 'pf-u-mt-lg',
    mtXl: 'pf-u-mt-xl',
    mt_2xl: 'pf-u-mt-2xl',
    mt_3xl: 'pf-u-mt-3xl',
    mrAuto: 'pf-u-mr-auto',
    mr_0: 'pf-u-mr-0',
    mrXs: 'pf-u-mr-xs',
    mrSm: 'pf-u-mr-sm',
    mrMd: 'pf-u-mr-md',
    mrLg: 'pf-u-mr-lg',
    mrXl: 'pf-u-mr-xl',
    mr_2xl: 'pf-u-mr-2xl',
    mr_3xl: 'pf-u-mr-3xl',
    mbAuto: 'pf-u-mb-auto',
    mb_0: 'pf-u-mb-0',
    mbXs: 'pf-u-mb-xs',
    mbSm: 'pf-u-mb-sm',
    mbMd: 'pf-u-mb-md',
    mbLg: 'pf-u-mb-lg',
    mbXl: 'pf-u-mb-xl',
    mb_2xl: 'pf-u-mb-2xl',
    mb_3xl: 'pf-u-mb-3xl',
    mlAuto: 'pf-u-ml-auto',
    ml_0: 'pf-u-ml-0',
    mlXs: 'pf-u-ml-xs',
    mlSm: 'pf-u-ml-sm',
    mlMd: 'pf-u-ml-md',
    mlLg: 'pf-u-ml-lg',
    mlXl: 'pf-u-ml-xl',
    ml_2xl: 'pf-u-ml-2xl',
    ml_3xl: 'pf-u-ml-3xl',
    mxAuto: 'pf-u-mx-auto',
    mx_0: 'pf-u-mx-0',
    mxXs: 'pf-u-mx-xs',
    mxSm: 'pf-u-mx-sm',
    mxMd: 'pf-u-mx-md',
    mxLg: 'pf-u-mx-lg',
    mxXl: 'pf-u-mx-xl',
    mx_2xl: 'pf-u-mx-2xl',
    mx_3xl: 'pf-u-mx-3xl',
    myAuto: 'pf-u-my-auto',
    my_0: 'pf-u-my-0',
    myXs: 'pf-u-my-xs',
    mySm: 'pf-u-my-sm',
    myMd: 'pf-u-my-md',
    myLg: 'pf-u-my-lg',
    myXl: 'pf-u-my-xl',
    my_2xl: 'pf-u-my-2xl',
    my_3xl: 'pf-u-my-3xl',
    pAuto: 'pf-u-p-auto',
    p_0: 'pf-u-p-0',
    pXs: 'pf-u-p-xs',
    pSm: 'pf-u-p-sm',
    pMd: 'pf-u-p-md',
    pLg: 'pf-u-p-lg',
    pXl: 'pf-u-p-xl',
    p_2xl: 'pf-u-p-2xl',
    p_3xl: 'pf-u-p-3xl',
    ptAuto: 'pf-u-pt-auto',
    pt_0: 'pf-u-pt-0',
    ptXs: 'pf-u-pt-xs',
    ptSm: 'pf-u-pt-sm',
    ptMd: 'pf-u-pt-md',
    ptLg: 'pf-u-pt-lg',
    ptXl: 'pf-u-pt-xl',
    pt_2xl: 'pf-u-pt-2xl',
    pt_3xl: 'pf-u-pt-3xl',
    prAuto: 'pf-u-pr-auto',
    pr_0: 'pf-u-pr-0',
    prXs: 'pf-u-pr-xs',
    prSm: 'pf-u-pr-sm',
    prMd: 'pf-u-pr-md',
    prLg: 'pf-u-pr-lg',
    prXl: 'pf-u-pr-xl',
    pr_2xl: 'pf-u-pr-2xl',
    pr_3xl: 'pf-u-pr-3xl',
    pbAuto: 'pf-u-pb-auto',
    pb_0: 'pf-u-pb-0',
    pbXs: 'pf-u-pb-xs',
    pbSm: 'pf-u-pb-sm',
    pbMd: 'pf-u-pb-md',
    pbLg: 'pf-u-pb-lg',
    pbXl: 'pf-u-pb-xl',
    pb_2xl: 'pf-u-pb-2xl',
    pb_3xl: 'pf-u-pb-3xl',
    plAuto: 'pf-u-pl-auto',
    pl_0: 'pf-u-pl-0',
    plXs: 'pf-u-pl-xs',
    plSm: 'pf-u-pl-sm',
    plMd: 'pf-u-pl-md',
    plLg: 'pf-u-pl-lg',
    plXl: 'pf-u-pl-xl',
    pl_2xl: 'pf-u-pl-2xl',
    pl_3xl: 'pf-u-pl-3xl',
    pxAuto: 'pf-u-px-auto',
    px_0: 'pf-u-px-0',
    pxXs: 'pf-u-px-xs',
    pxSm: 'pf-u-px-sm',
    pxMd: 'pf-u-px-md',
    pxLg: 'pf-u-px-lg',
    pxXl: 'pf-u-px-xl',
    px_2xl: 'pf-u-px-2xl',
    px_3xl: 'pf-u-px-3xl',
    pyAuto: 'pf-u-py-auto',
    py_0: 'pf-u-py-0',
    pyXs: 'pf-u-py-xs',
    pySm: 'pf-u-py-sm',
    pyMd: 'pf-u-py-md',
    pyLg: 'pf-u-py-lg',
    pyXl: 'pf-u-py-xl',
    py_2xl: 'pf-u-py-2xl',
    py_3xl: 'pf-u-py-3xl',
    mAutoOnSm: 'pf-u-m-auto-on-sm',
    m_0OnSm: 'pf-u-m-0-on-sm',
    mXsOnSm: 'pf-u-m-xs-on-sm',
    mSmOnSm: 'pf-u-m-sm-on-sm',
    mMdOnSm: 'pf-u-m-md-on-sm',
    mLgOnSm: 'pf-u-m-lg-on-sm',
    mXlOnSm: 'pf-u-m-xl-on-sm',
    m_2xlOnSm: 'pf-u-m-2xl-on-sm',
    m_3xlOnSm: 'pf-u-m-3xl-on-sm',
    mtAutoOnSm: 'pf-u-mt-auto-on-sm',
    mt_0OnSm: 'pf-u-mt-0-on-sm',
    mtXsOnSm: 'pf-u-mt-xs-on-sm',
    mtSmOnSm: 'pf-u-mt-sm-on-sm',
    mtMdOnSm: 'pf-u-mt-md-on-sm',
    mtLgOnSm: 'pf-u-mt-lg-on-sm',
    mtXlOnSm: 'pf-u-mt-xl-on-sm',
    mt_2xlOnSm: 'pf-u-mt-2xl-on-sm',
    mt_3xlOnSm: 'pf-u-mt-3xl-on-sm',
    mrAutoOnSm: 'pf-u-mr-auto-on-sm',
    mr_0OnSm: 'pf-u-mr-0-on-sm',
    mrXsOnSm: 'pf-u-mr-xs-on-sm',
    mrSmOnSm: 'pf-u-mr-sm-on-sm',
    mrMdOnSm: 'pf-u-mr-md-on-sm',
    mrLgOnSm: 'pf-u-mr-lg-on-sm',
    mrXlOnSm: 'pf-u-mr-xl-on-sm',
    mr_2xlOnSm: 'pf-u-mr-2xl-on-sm',
    mr_3xlOnSm: 'pf-u-mr-3xl-on-sm',
    mbAutoOnSm: 'pf-u-mb-auto-on-sm',
    mb_0OnSm: 'pf-u-mb-0-on-sm',
    mbXsOnSm: 'pf-u-mb-xs-on-sm',
    mbSmOnSm: 'pf-u-mb-sm-on-sm',
    mbMdOnSm: 'pf-u-mb-md-on-sm',
    mbLgOnSm: 'pf-u-mb-lg-on-sm',
    mbXlOnSm: 'pf-u-mb-xl-on-sm',
    mb_2xlOnSm: 'pf-u-mb-2xl-on-sm',
    mb_3xlOnSm: 'pf-u-mb-3xl-on-sm',
    mlAutoOnSm: 'pf-u-ml-auto-on-sm',
    ml_0OnSm: 'pf-u-ml-0-on-sm',
    mlXsOnSm: 'pf-u-ml-xs-on-sm',
    mlSmOnSm: 'pf-u-ml-sm-on-sm',
    mlMdOnSm: 'pf-u-ml-md-on-sm',
    mlLgOnSm: 'pf-u-ml-lg-on-sm',
    mlXlOnSm: 'pf-u-ml-xl-on-sm',
    ml_2xlOnSm: 'pf-u-ml-2xl-on-sm',
    ml_3xlOnSm: 'pf-u-ml-3xl-on-sm',
    mxAutoOnSm: 'pf-u-mx-auto-on-sm',
    mx_0OnSm: 'pf-u-mx-0-on-sm',
    mxXsOnSm: 'pf-u-mx-xs-on-sm',
    mxSmOnSm: 'pf-u-mx-sm-on-sm',
    mxMdOnSm: 'pf-u-mx-md-on-sm',
    mxLgOnSm: 'pf-u-mx-lg-on-sm',
    mxXlOnSm: 'pf-u-mx-xl-on-sm',
    mx_2xlOnSm: 'pf-u-mx-2xl-on-sm',
    mx_3xlOnSm: 'pf-u-mx-3xl-on-sm',
    myAutoOnSm: 'pf-u-my-auto-on-sm',
    my_0OnSm: 'pf-u-my-0-on-sm',
    myXsOnSm: 'pf-u-my-xs-on-sm',
    mySmOnSm: 'pf-u-my-sm-on-sm',
    myMdOnSm: 'pf-u-my-md-on-sm',
    myLgOnSm: 'pf-u-my-lg-on-sm',
    myXlOnSm: 'pf-u-my-xl-on-sm',
    my_2xlOnSm: 'pf-u-my-2xl-on-sm',
    my_3xlOnSm: 'pf-u-my-3xl-on-sm',
    pAutoOnSm: 'pf-u-p-auto-on-sm',
    p_0OnSm: 'pf-u-p-0-on-sm',
    pXsOnSm: 'pf-u-p-xs-on-sm',
    pSmOnSm: 'pf-u-p-sm-on-sm',
    pMdOnSm: 'pf-u-p-md-on-sm',
    pLgOnSm: 'pf-u-p-lg-on-sm',
    pXlOnSm: 'pf-u-p-xl-on-sm',
    p_2xlOnSm: 'pf-u-p-2xl-on-sm',
    p_3xlOnSm: 'pf-u-p-3xl-on-sm',
    ptAutoOnSm: 'pf-u-pt-auto-on-sm',
    pt_0OnSm: 'pf-u-pt-0-on-sm',
    ptXsOnSm: 'pf-u-pt-xs-on-sm',
    ptSmOnSm: 'pf-u-pt-sm-on-sm',
    ptMdOnSm: 'pf-u-pt-md-on-sm',
    ptLgOnSm: 'pf-u-pt-lg-on-sm',
    ptXlOnSm: 'pf-u-pt-xl-on-sm',
    pt_2xlOnSm: 'pf-u-pt-2xl-on-sm',
    pt_3xlOnSm: 'pf-u-pt-3xl-on-sm',
    prAutoOnSm: 'pf-u-pr-auto-on-sm',
    pr_0OnSm: 'pf-u-pr-0-on-sm',
    prXsOnSm: 'pf-u-pr-xs-on-sm',
    prSmOnSm: 'pf-u-pr-sm-on-sm',
    prMdOnSm: 'pf-u-pr-md-on-sm',
    prLgOnSm: 'pf-u-pr-lg-on-sm',
    prXlOnSm: 'pf-u-pr-xl-on-sm',
    pr_2xlOnSm: 'pf-u-pr-2xl-on-sm',
    pr_3xlOnSm: 'pf-u-pr-3xl-on-sm',
    pbAutoOnSm: 'pf-u-pb-auto-on-sm',
    pb_0OnSm: 'pf-u-pb-0-on-sm',
    pbXsOnSm: 'pf-u-pb-xs-on-sm',
    pbSmOnSm: 'pf-u-pb-sm-on-sm',
    pbMdOnSm: 'pf-u-pb-md-on-sm',
    pbLgOnSm: 'pf-u-pb-lg-on-sm',
    pbXlOnSm: 'pf-u-pb-xl-on-sm',
    pb_2xlOnSm: 'pf-u-pb-2xl-on-sm',
    pb_3xlOnSm: 'pf-u-pb-3xl-on-sm',
    plAutoOnSm: 'pf-u-pl-auto-on-sm',
    pl_0OnSm: 'pf-u-pl-0-on-sm',
    plXsOnSm: 'pf-u-pl-xs-on-sm',
    plSmOnSm: 'pf-u-pl-sm-on-sm',
    plMdOnSm: 'pf-u-pl-md-on-sm',
    plLgOnSm: 'pf-u-pl-lg-on-sm',
    plXlOnSm: 'pf-u-pl-xl-on-sm',
    pl_2xlOnSm: 'pf-u-pl-2xl-on-sm',
    pl_3xlOnSm: 'pf-u-pl-3xl-on-sm',
    pxAutoOnSm: 'pf-u-px-auto-on-sm',
    px_0OnSm: 'pf-u-px-0-on-sm',
    pxXsOnSm: 'pf-u-px-xs-on-sm',
    pxSmOnSm: 'pf-u-px-sm-on-sm',
    pxMdOnSm: 'pf-u-px-md-on-sm',
    pxLgOnSm: 'pf-u-px-lg-on-sm',
    pxXlOnSm: 'pf-u-px-xl-on-sm',
    px_2xlOnSm: 'pf-u-px-2xl-on-sm',
    px_3xlOnSm: 'pf-u-px-3xl-on-sm',
    pyAutoOnSm: 'pf-u-py-auto-on-sm',
    py_0OnSm: 'pf-u-py-0-on-sm',
    pyXsOnSm: 'pf-u-py-xs-on-sm',
    pySmOnSm: 'pf-u-py-sm-on-sm',
    pyMdOnSm: 'pf-u-py-md-on-sm',
    pyLgOnSm: 'pf-u-py-lg-on-sm',
    pyXlOnSm: 'pf-u-py-xl-on-sm',
    py_2xlOnSm: 'pf-u-py-2xl-on-sm',
    py_3xlOnSm: 'pf-u-py-3xl-on-sm',
    mAutoOnMd: 'pf-u-m-auto-on-md',
    m_0OnMd: 'pf-u-m-0-on-md',
    mXsOnMd: 'pf-u-m-xs-on-md',
    mSmOnMd: 'pf-u-m-sm-on-md',
    mMdOnMd: 'pf-u-m-md-on-md',
    mLgOnMd: 'pf-u-m-lg-on-md',
    mXlOnMd: 'pf-u-m-xl-on-md',
    m_2xlOnMd: 'pf-u-m-2xl-on-md',
    m_3xlOnMd: 'pf-u-m-3xl-on-md',
    mtAutoOnMd: 'pf-u-mt-auto-on-md',
    mt_0OnMd: 'pf-u-mt-0-on-md',
    mtXsOnMd: 'pf-u-mt-xs-on-md',
    mtSmOnMd: 'pf-u-mt-sm-on-md',
    mtMdOnMd: 'pf-u-mt-md-on-md',
    mtLgOnMd: 'pf-u-mt-lg-on-md',
    mtXlOnMd: 'pf-u-mt-xl-on-md',
    mt_2xlOnMd: 'pf-u-mt-2xl-on-md',
    mt_3xlOnMd: 'pf-u-mt-3xl-on-md',
    mrAutoOnMd: 'pf-u-mr-auto-on-md',
    mr_0OnMd: 'pf-u-mr-0-on-md',
    mrXsOnMd: 'pf-u-mr-xs-on-md',
    mrSmOnMd: 'pf-u-mr-sm-on-md',
    mrMdOnMd: 'pf-u-mr-md-on-md',
    mrLgOnMd: 'pf-u-mr-lg-on-md',
    mrXlOnMd: 'pf-u-mr-xl-on-md',
    mr_2xlOnMd: 'pf-u-mr-2xl-on-md',
    mr_3xlOnMd: 'pf-u-mr-3xl-on-md',
    mbAutoOnMd: 'pf-u-mb-auto-on-md',
    mb_0OnMd: 'pf-u-mb-0-on-md',
    mbXsOnMd: 'pf-u-mb-xs-on-md',
    mbSmOnMd: 'pf-u-mb-sm-on-md',
    mbMdOnMd: 'pf-u-mb-md-on-md',
    mbLgOnMd: 'pf-u-mb-lg-on-md',
    mbXlOnMd: 'pf-u-mb-xl-on-md',
    mb_2xlOnMd: 'pf-u-mb-2xl-on-md',
    mb_3xlOnMd: 'pf-u-mb-3xl-on-md',
    mlAutoOnMd: 'pf-u-ml-auto-on-md',
    ml_0OnMd: 'pf-u-ml-0-on-md',
    mlXsOnMd: 'pf-u-ml-xs-on-md',
    mlSmOnMd: 'pf-u-ml-sm-on-md',
    mlMdOnMd: 'pf-u-ml-md-on-md',
    mlLgOnMd: 'pf-u-ml-lg-on-md',
    mlXlOnMd: 'pf-u-ml-xl-on-md',
    ml_2xlOnMd: 'pf-u-ml-2xl-on-md',
    ml_3xlOnMd: 'pf-u-ml-3xl-on-md',
    mxAutoOnMd: 'pf-u-mx-auto-on-md',
    mx_0OnMd: 'pf-u-mx-0-on-md',
    mxXsOnMd: 'pf-u-mx-xs-on-md',
    mxSmOnMd: 'pf-u-mx-sm-on-md',
    mxMdOnMd: 'pf-u-mx-md-on-md',
    mxLgOnMd: 'pf-u-mx-lg-on-md',
    mxXlOnMd: 'pf-u-mx-xl-on-md',
    mx_2xlOnMd: 'pf-u-mx-2xl-on-md',
    mx_3xlOnMd: 'pf-u-mx-3xl-on-md',
    myAutoOnMd: 'pf-u-my-auto-on-md',
    my_0OnMd: 'pf-u-my-0-on-md',
    myXsOnMd: 'pf-u-my-xs-on-md',
    mySmOnMd: 'pf-u-my-sm-on-md',
    myMdOnMd: 'pf-u-my-md-on-md',
    myLgOnMd: 'pf-u-my-lg-on-md',
    myXlOnMd: 'pf-u-my-xl-on-md',
    my_2xlOnMd: 'pf-u-my-2xl-on-md',
    my_3xlOnMd: 'pf-u-my-3xl-on-md',
    pAutoOnMd: 'pf-u-p-auto-on-md',
    p_0OnMd: 'pf-u-p-0-on-md',
    pXsOnMd: 'pf-u-p-xs-on-md',
    pSmOnMd: 'pf-u-p-sm-on-md',
    pMdOnMd: 'pf-u-p-md-on-md',
    pLgOnMd: 'pf-u-p-lg-on-md',
    pXlOnMd: 'pf-u-p-xl-on-md',
    p_2xlOnMd: 'pf-u-p-2xl-on-md',
    p_3xlOnMd: 'pf-u-p-3xl-on-md',
    ptAutoOnMd: 'pf-u-pt-auto-on-md',
    pt_0OnMd: 'pf-u-pt-0-on-md',
    ptXsOnMd: 'pf-u-pt-xs-on-md',
    ptSmOnMd: 'pf-u-pt-sm-on-md',
    ptMdOnMd: 'pf-u-pt-md-on-md',
    ptLgOnMd: 'pf-u-pt-lg-on-md',
    ptXlOnMd: 'pf-u-pt-xl-on-md',
    pt_2xlOnMd: 'pf-u-pt-2xl-on-md',
    pt_3xlOnMd: 'pf-u-pt-3xl-on-md',
    prAutoOnMd: 'pf-u-pr-auto-on-md',
    pr_0OnMd: 'pf-u-pr-0-on-md',
    prXsOnMd: 'pf-u-pr-xs-on-md',
    prSmOnMd: 'pf-u-pr-sm-on-md',
    prMdOnMd: 'pf-u-pr-md-on-md',
    prLgOnMd: 'pf-u-pr-lg-on-md',
    prXlOnMd: 'pf-u-pr-xl-on-md',
    pr_2xlOnMd: 'pf-u-pr-2xl-on-md',
    pr_3xlOnMd: 'pf-u-pr-3xl-on-md',
    pbAutoOnMd: 'pf-u-pb-auto-on-md',
    pb_0OnMd: 'pf-u-pb-0-on-md',
    pbXsOnMd: 'pf-u-pb-xs-on-md',
    pbSmOnMd: 'pf-u-pb-sm-on-md',
    pbMdOnMd: 'pf-u-pb-md-on-md',
    pbLgOnMd: 'pf-u-pb-lg-on-md',
    pbXlOnMd: 'pf-u-pb-xl-on-md',
    pb_2xlOnMd: 'pf-u-pb-2xl-on-md',
    pb_3xlOnMd: 'pf-u-pb-3xl-on-md',
    plAutoOnMd: 'pf-u-pl-auto-on-md',
    pl_0OnMd: 'pf-u-pl-0-on-md',
    plXsOnMd: 'pf-u-pl-xs-on-md',
    plSmOnMd: 'pf-u-pl-sm-on-md',
    plMdOnMd: 'pf-u-pl-md-on-md',
    plLgOnMd: 'pf-u-pl-lg-on-md',
    plXlOnMd: 'pf-u-pl-xl-on-md',
    pl_2xlOnMd: 'pf-u-pl-2xl-on-md',
    pl_3xlOnMd: 'pf-u-pl-3xl-on-md',
    pxAutoOnMd: 'pf-u-px-auto-on-md',
    px_0OnMd: 'pf-u-px-0-on-md',
    pxXsOnMd: 'pf-u-px-xs-on-md',
    pxSmOnMd: 'pf-u-px-sm-on-md',
    pxMdOnMd: 'pf-u-px-md-on-md',
    pxLgOnMd: 'pf-u-px-lg-on-md',
    pxXlOnMd: 'pf-u-px-xl-on-md',
    px_2xlOnMd: 'pf-u-px-2xl-on-md',
    px_3xlOnMd: 'pf-u-px-3xl-on-md',
    pyAutoOnMd: 'pf-u-py-auto-on-md',
    py_0OnMd: 'pf-u-py-0-on-md',
    pyXsOnMd: 'pf-u-py-xs-on-md',
    pySmOnMd: 'pf-u-py-sm-on-md',
    pyMdOnMd: 'pf-u-py-md-on-md',
    pyLgOnMd: 'pf-u-py-lg-on-md',
    pyXlOnMd: 'pf-u-py-xl-on-md',
    py_2xlOnMd: 'pf-u-py-2xl-on-md',
    py_3xlOnMd: 'pf-u-py-3xl-on-md',
    mAutoOnLg: 'pf-u-m-auto-on-lg',
    m_0OnLg: 'pf-u-m-0-on-lg',
    mXsOnLg: 'pf-u-m-xs-on-lg',
    mSmOnLg: 'pf-u-m-sm-on-lg',
    mMdOnLg: 'pf-u-m-md-on-lg',
    mLgOnLg: 'pf-u-m-lg-on-lg',
    mXlOnLg: 'pf-u-m-xl-on-lg',
    m_2xlOnLg: 'pf-u-m-2xl-on-lg',
    m_3xlOnLg: 'pf-u-m-3xl-on-lg',
    mtAutoOnLg: 'pf-u-mt-auto-on-lg',
    mt_0OnLg: 'pf-u-mt-0-on-lg',
    mtXsOnLg: 'pf-u-mt-xs-on-lg',
    mtSmOnLg: 'pf-u-mt-sm-on-lg',
    mtMdOnLg: 'pf-u-mt-md-on-lg',
    mtLgOnLg: 'pf-u-mt-lg-on-lg',
    mtXlOnLg: 'pf-u-mt-xl-on-lg',
    mt_2xlOnLg: 'pf-u-mt-2xl-on-lg',
    mt_3xlOnLg: 'pf-u-mt-3xl-on-lg',
    mrAutoOnLg: 'pf-u-mr-auto-on-lg',
    mr_0OnLg: 'pf-u-mr-0-on-lg',
    mrXsOnLg: 'pf-u-mr-xs-on-lg',
    mrSmOnLg: 'pf-u-mr-sm-on-lg',
    mrMdOnLg: 'pf-u-mr-md-on-lg',
    mrLgOnLg: 'pf-u-mr-lg-on-lg',
    mrXlOnLg: 'pf-u-mr-xl-on-lg',
    mr_2xlOnLg: 'pf-u-mr-2xl-on-lg',
    mr_3xlOnLg: 'pf-u-mr-3xl-on-lg',
    mbAutoOnLg: 'pf-u-mb-auto-on-lg',
    mb_0OnLg: 'pf-u-mb-0-on-lg',
    mbXsOnLg: 'pf-u-mb-xs-on-lg',
    mbSmOnLg: 'pf-u-mb-sm-on-lg',
    mbMdOnLg: 'pf-u-mb-md-on-lg',
    mbLgOnLg: 'pf-u-mb-lg-on-lg',
    mbXlOnLg: 'pf-u-mb-xl-on-lg',
    mb_2xlOnLg: 'pf-u-mb-2xl-on-lg',
    mb_3xlOnLg: 'pf-u-mb-3xl-on-lg',
    mlAutoOnLg: 'pf-u-ml-auto-on-lg',
    ml_0OnLg: 'pf-u-ml-0-on-lg',
    mlXsOnLg: 'pf-u-ml-xs-on-lg',
    mlSmOnLg: 'pf-u-ml-sm-on-lg',
    mlMdOnLg: 'pf-u-ml-md-on-lg',
    mlLgOnLg: 'pf-u-ml-lg-on-lg',
    mlXlOnLg: 'pf-u-ml-xl-on-lg',
    ml_2xlOnLg: 'pf-u-ml-2xl-on-lg',
    ml_3xlOnLg: 'pf-u-ml-3xl-on-lg',
    mxAutoOnLg: 'pf-u-mx-auto-on-lg',
    mx_0OnLg: 'pf-u-mx-0-on-lg',
    mxXsOnLg: 'pf-u-mx-xs-on-lg',
    mxSmOnLg: 'pf-u-mx-sm-on-lg',
    mxMdOnLg: 'pf-u-mx-md-on-lg',
    mxLgOnLg: 'pf-u-mx-lg-on-lg',
    mxXlOnLg: 'pf-u-mx-xl-on-lg',
    mx_2xlOnLg: 'pf-u-mx-2xl-on-lg',
    mx_3xlOnLg: 'pf-u-mx-3xl-on-lg',
    myAutoOnLg: 'pf-u-my-auto-on-lg',
    my_0OnLg: 'pf-u-my-0-on-lg',
    myXsOnLg: 'pf-u-my-xs-on-lg',
    mySmOnLg: 'pf-u-my-sm-on-lg',
    myMdOnLg: 'pf-u-my-md-on-lg',
    myLgOnLg: 'pf-u-my-lg-on-lg',
    myXlOnLg: 'pf-u-my-xl-on-lg',
    my_2xlOnLg: 'pf-u-my-2xl-on-lg',
    my_3xlOnLg: 'pf-u-my-3xl-on-lg',
    pAutoOnLg: 'pf-u-p-auto-on-lg',
    p_0OnLg: 'pf-u-p-0-on-lg',
    pXsOnLg: 'pf-u-p-xs-on-lg',
    pSmOnLg: 'pf-u-p-sm-on-lg',
    pMdOnLg: 'pf-u-p-md-on-lg',
    pLgOnLg: 'pf-u-p-lg-on-lg',
    pXlOnLg: 'pf-u-p-xl-on-lg',
    p_2xlOnLg: 'pf-u-p-2xl-on-lg',
    p_3xlOnLg: 'pf-u-p-3xl-on-lg',
    ptAutoOnLg: 'pf-u-pt-auto-on-lg',
    pt_0OnLg: 'pf-u-pt-0-on-lg',
    ptXsOnLg: 'pf-u-pt-xs-on-lg',
    ptSmOnLg: 'pf-u-pt-sm-on-lg',
    ptMdOnLg: 'pf-u-pt-md-on-lg',
    ptLgOnLg: 'pf-u-pt-lg-on-lg',
    ptXlOnLg: 'pf-u-pt-xl-on-lg',
    pt_2xlOnLg: 'pf-u-pt-2xl-on-lg',
    pt_3xlOnLg: 'pf-u-pt-3xl-on-lg',
    prAutoOnLg: 'pf-u-pr-auto-on-lg',
    pr_0OnLg: 'pf-u-pr-0-on-lg',
    prXsOnLg: 'pf-u-pr-xs-on-lg',
    prSmOnLg: 'pf-u-pr-sm-on-lg',
    prMdOnLg: 'pf-u-pr-md-on-lg',
    prLgOnLg: 'pf-u-pr-lg-on-lg',
    prXlOnLg: 'pf-u-pr-xl-on-lg',
    pr_2xlOnLg: 'pf-u-pr-2xl-on-lg',
    pr_3xlOnLg: 'pf-u-pr-3xl-on-lg',
    pbAutoOnLg: 'pf-u-pb-auto-on-lg',
    pb_0OnLg: 'pf-u-pb-0-on-lg',
    pbXsOnLg: 'pf-u-pb-xs-on-lg',
    pbSmOnLg: 'pf-u-pb-sm-on-lg',
    pbMdOnLg: 'pf-u-pb-md-on-lg',
    pbLgOnLg: 'pf-u-pb-lg-on-lg',
    pbXlOnLg: 'pf-u-pb-xl-on-lg',
    pb_2xlOnLg: 'pf-u-pb-2xl-on-lg',
    pb_3xlOnLg: 'pf-u-pb-3xl-on-lg',
    plAutoOnLg: 'pf-u-pl-auto-on-lg',
    pl_0OnLg: 'pf-u-pl-0-on-lg',
    plXsOnLg: 'pf-u-pl-xs-on-lg',
    plSmOnLg: 'pf-u-pl-sm-on-lg',
    plMdOnLg: 'pf-u-pl-md-on-lg',
    plLgOnLg: 'pf-u-pl-lg-on-lg',
    plXlOnLg: 'pf-u-pl-xl-on-lg',
    pl_2xlOnLg: 'pf-u-pl-2xl-on-lg',
    pl_3xlOnLg: 'pf-u-pl-3xl-on-lg',
    pxAutoOnLg: 'pf-u-px-auto-on-lg',
    px_0OnLg: 'pf-u-px-0-on-lg',
    pxXsOnLg: 'pf-u-px-xs-on-lg',
    pxSmOnLg: 'pf-u-px-sm-on-lg',
    pxMdOnLg: 'pf-u-px-md-on-lg',
    pxLgOnLg: 'pf-u-px-lg-on-lg',
    pxXlOnLg: 'pf-u-px-xl-on-lg',
    px_2xlOnLg: 'pf-u-px-2xl-on-lg',
    px_3xlOnLg: 'pf-u-px-3xl-on-lg',
    pyAutoOnLg: 'pf-u-py-auto-on-lg',
    py_0OnLg: 'pf-u-py-0-on-lg',
    pyXsOnLg: 'pf-u-py-xs-on-lg',
    pySmOnLg: 'pf-u-py-sm-on-lg',
    pyMdOnLg: 'pf-u-py-md-on-lg',
    pyLgOnLg: 'pf-u-py-lg-on-lg',
    pyXlOnLg: 'pf-u-py-xl-on-lg',
    py_2xlOnLg: 'pf-u-py-2xl-on-lg',
    py_3xlOnLg: 'pf-u-py-3xl-on-lg',
    mAutoOnXl: 'pf-u-m-auto-on-xl',
    m_0OnXl: 'pf-u-m-0-on-xl',
    mXsOnXl: 'pf-u-m-xs-on-xl',
    mSmOnXl: 'pf-u-m-sm-on-xl',
    mMdOnXl: 'pf-u-m-md-on-xl',
    mLgOnXl: 'pf-u-m-lg-on-xl',
    mXlOnXl: 'pf-u-m-xl-on-xl',
    m_2xlOnXl: 'pf-u-m-2xl-on-xl',
    m_3xlOnXl: 'pf-u-m-3xl-on-xl',
    mtAutoOnXl: 'pf-u-mt-auto-on-xl',
    mt_0OnXl: 'pf-u-mt-0-on-xl',
    mtXsOnXl: 'pf-u-mt-xs-on-xl',
    mtSmOnXl: 'pf-u-mt-sm-on-xl',
    mtMdOnXl: 'pf-u-mt-md-on-xl',
    mtLgOnXl: 'pf-u-mt-lg-on-xl',
    mtXlOnXl: 'pf-u-mt-xl-on-xl',
    mt_2xlOnXl: 'pf-u-mt-2xl-on-xl',
    mt_3xlOnXl: 'pf-u-mt-3xl-on-xl',
    mrAutoOnXl: 'pf-u-mr-auto-on-xl',
    mr_0OnXl: 'pf-u-mr-0-on-xl',
    mrXsOnXl: 'pf-u-mr-xs-on-xl',
    mrSmOnXl: 'pf-u-mr-sm-on-xl',
    mrMdOnXl: 'pf-u-mr-md-on-xl',
    mrLgOnXl: 'pf-u-mr-lg-on-xl',
    mrXlOnXl: 'pf-u-mr-xl-on-xl',
    mr_2xlOnXl: 'pf-u-mr-2xl-on-xl',
    mr_3xlOnXl: 'pf-u-mr-3xl-on-xl',
    mbAutoOnXl: 'pf-u-mb-auto-on-xl',
    mb_0OnXl: 'pf-u-mb-0-on-xl',
    mbXsOnXl: 'pf-u-mb-xs-on-xl',
    mbSmOnXl: 'pf-u-mb-sm-on-xl',
    mbMdOnXl: 'pf-u-mb-md-on-xl',
    mbLgOnXl: 'pf-u-mb-lg-on-xl',
    mbXlOnXl: 'pf-u-mb-xl-on-xl',
    mb_2xlOnXl: 'pf-u-mb-2xl-on-xl',
    mb_3xlOnXl: 'pf-u-mb-3xl-on-xl',
    mlAutoOnXl: 'pf-u-ml-auto-on-xl',
    ml_0OnXl: 'pf-u-ml-0-on-xl',
    mlXsOnXl: 'pf-u-ml-xs-on-xl',
    mlSmOnXl: 'pf-u-ml-sm-on-xl',
    mlMdOnXl: 'pf-u-ml-md-on-xl',
    mlLgOnXl: 'pf-u-ml-lg-on-xl',
    mlXlOnXl: 'pf-u-ml-xl-on-xl',
    ml_2xlOnXl: 'pf-u-ml-2xl-on-xl',
    ml_3xlOnXl: 'pf-u-ml-3xl-on-xl',
    mxAutoOnXl: 'pf-u-mx-auto-on-xl',
    mx_0OnXl: 'pf-u-mx-0-on-xl',
    mxXsOnXl: 'pf-u-mx-xs-on-xl',
    mxSmOnXl: 'pf-u-mx-sm-on-xl',
    mxMdOnXl: 'pf-u-mx-md-on-xl',
    mxLgOnXl: 'pf-u-mx-lg-on-xl',
    mxXlOnXl: 'pf-u-mx-xl-on-xl',
    mx_2xlOnXl: 'pf-u-mx-2xl-on-xl',
    mx_3xlOnXl: 'pf-u-mx-3xl-on-xl',
    myAutoOnXl: 'pf-u-my-auto-on-xl',
    my_0OnXl: 'pf-u-my-0-on-xl',
    myXsOnXl: 'pf-u-my-xs-on-xl',
    mySmOnXl: 'pf-u-my-sm-on-xl',
    myMdOnXl: 'pf-u-my-md-on-xl',
    myLgOnXl: 'pf-u-my-lg-on-xl',
    myXlOnXl: 'pf-u-my-xl-on-xl',
    my_2xlOnXl: 'pf-u-my-2xl-on-xl',
    my_3xlOnXl: 'pf-u-my-3xl-on-xl',
    pAutoOnXl: 'pf-u-p-auto-on-xl',
    p_0OnXl: 'pf-u-p-0-on-xl',
    pXsOnXl: 'pf-u-p-xs-on-xl',
    pSmOnXl: 'pf-u-p-sm-on-xl',
    pMdOnXl: 'pf-u-p-md-on-xl',
    pLgOnXl: 'pf-u-p-lg-on-xl',
    pXlOnXl: 'pf-u-p-xl-on-xl',
    p_2xlOnXl: 'pf-u-p-2xl-on-xl',
    p_3xlOnXl: 'pf-u-p-3xl-on-xl',
    ptAutoOnXl: 'pf-u-pt-auto-on-xl',
    pt_0OnXl: 'pf-u-pt-0-on-xl',
    ptXsOnXl: 'pf-u-pt-xs-on-xl',
    ptSmOnXl: 'pf-u-pt-sm-on-xl',
    ptMdOnXl: 'pf-u-pt-md-on-xl',
    ptLgOnXl: 'pf-u-pt-lg-on-xl',
    ptXlOnXl: 'pf-u-pt-xl-on-xl',
    pt_2xlOnXl: 'pf-u-pt-2xl-on-xl',
    pt_3xlOnXl: 'pf-u-pt-3xl-on-xl',
    prAutoOnXl: 'pf-u-pr-auto-on-xl',
    pr_0OnXl: 'pf-u-pr-0-on-xl',
    prXsOnXl: 'pf-u-pr-xs-on-xl',
    prSmOnXl: 'pf-u-pr-sm-on-xl',
    prMdOnXl: 'pf-u-pr-md-on-xl',
    prLgOnXl: 'pf-u-pr-lg-on-xl',
    prXlOnXl: 'pf-u-pr-xl-on-xl',
    pr_2xlOnXl: 'pf-u-pr-2xl-on-xl',
    pr_3xlOnXl: 'pf-u-pr-3xl-on-xl',
    pbAutoOnXl: 'pf-u-pb-auto-on-xl',
    pb_0OnXl: 'pf-u-pb-0-on-xl',
    pbXsOnXl: 'pf-u-pb-xs-on-xl',
    pbSmOnXl: 'pf-u-pb-sm-on-xl',
    pbMdOnXl: 'pf-u-pb-md-on-xl',
    pbLgOnXl: 'pf-u-pb-lg-on-xl',
    pbXlOnXl: 'pf-u-pb-xl-on-xl',
    pb_2xlOnXl: 'pf-u-pb-2xl-on-xl',
    pb_3xlOnXl: 'pf-u-pb-3xl-on-xl',
    plAutoOnXl: 'pf-u-pl-auto-on-xl',
    pl_0OnXl: 'pf-u-pl-0-on-xl',
    plXsOnXl: 'pf-u-pl-xs-on-xl',
    plSmOnXl: 'pf-u-pl-sm-on-xl',
    plMdOnXl: 'pf-u-pl-md-on-xl',
    plLgOnXl: 'pf-u-pl-lg-on-xl',
    plXlOnXl: 'pf-u-pl-xl-on-xl',
    pl_2xlOnXl: 'pf-u-pl-2xl-on-xl',
    pl_3xlOnXl: 'pf-u-pl-3xl-on-xl',
    pxAutoOnXl: 'pf-u-px-auto-on-xl',
    px_0OnXl: 'pf-u-px-0-on-xl',
    pxXsOnXl: 'pf-u-px-xs-on-xl',
    pxSmOnXl: 'pf-u-px-sm-on-xl',
    pxMdOnXl: 'pf-u-px-md-on-xl',
    pxLgOnXl: 'pf-u-px-lg-on-xl',
    pxXlOnXl: 'pf-u-px-xl-on-xl',
    px_2xlOnXl: 'pf-u-px-2xl-on-xl',
    px_3xlOnXl: 'pf-u-px-3xl-on-xl',
    pyAutoOnXl: 'pf-u-py-auto-on-xl',
    py_0OnXl: 'pf-u-py-0-on-xl',
    pyXsOnXl: 'pf-u-py-xs-on-xl',
    pySmOnXl: 'pf-u-py-sm-on-xl',
    pyMdOnXl: 'pf-u-py-md-on-xl',
    pyLgOnXl: 'pf-u-py-lg-on-xl',
    pyXlOnXl: 'pf-u-py-xl-on-xl',
    py_2xlOnXl: 'pf-u-py-2xl-on-xl',
    py_3xlOnXl: 'pf-u-py-3xl-on-xl',
    mAutoOn_2xl: 'pf-u-m-auto-on-2xl',
    m_0On_2xl: 'pf-u-m-0-on-2xl',
    mXsOn_2xl: 'pf-u-m-xs-on-2xl',
    mSmOn_2xl: 'pf-u-m-sm-on-2xl',
    mMdOn_2xl: 'pf-u-m-md-on-2xl',
    mLgOn_2xl: 'pf-u-m-lg-on-2xl',
    mXlOn_2xl: 'pf-u-m-xl-on-2xl',
    m_2xlOn_2xl: 'pf-u-m-2xl-on-2xl',
    m_3xlOn_2xl: 'pf-u-m-3xl-on-2xl',
    mtAutoOn_2xl: 'pf-u-mt-auto-on-2xl',
    mt_0On_2xl: 'pf-u-mt-0-on-2xl',
    mtXsOn_2xl: 'pf-u-mt-xs-on-2xl',
    mtSmOn_2xl: 'pf-u-mt-sm-on-2xl',
    mtMdOn_2xl: 'pf-u-mt-md-on-2xl',
    mtLgOn_2xl: 'pf-u-mt-lg-on-2xl',
    mtXlOn_2xl: 'pf-u-mt-xl-on-2xl',
    mt_2xlOn_2xl: 'pf-u-mt-2xl-on-2xl',
    mt_3xlOn_2xl: 'pf-u-mt-3xl-on-2xl',
    mrAutoOn_2xl: 'pf-u-mr-auto-on-2xl',
    mr_0On_2xl: 'pf-u-mr-0-on-2xl',
    mrXsOn_2xl: 'pf-u-mr-xs-on-2xl',
    mrSmOn_2xl: 'pf-u-mr-sm-on-2xl',
    mrMdOn_2xl: 'pf-u-mr-md-on-2xl',
    mrLgOn_2xl: 'pf-u-mr-lg-on-2xl',
    mrXlOn_2xl: 'pf-u-mr-xl-on-2xl',
    mr_2xlOn_2xl: 'pf-u-mr-2xl-on-2xl',
    mr_3xlOn_2xl: 'pf-u-mr-3xl-on-2xl',
    mbAutoOn_2xl: 'pf-u-mb-auto-on-2xl',
    mb_0On_2xl: 'pf-u-mb-0-on-2xl',
    mbXsOn_2xl: 'pf-u-mb-xs-on-2xl',
    mbSmOn_2xl: 'pf-u-mb-sm-on-2xl',
    mbMdOn_2xl: 'pf-u-mb-md-on-2xl',
    mbLgOn_2xl: 'pf-u-mb-lg-on-2xl',
    mbXlOn_2xl: 'pf-u-mb-xl-on-2xl',
    mb_2xlOn_2xl: 'pf-u-mb-2xl-on-2xl',
    mb_3xlOn_2xl: 'pf-u-mb-3xl-on-2xl',
    mlAutoOn_2xl: 'pf-u-ml-auto-on-2xl',
    ml_0On_2xl: 'pf-u-ml-0-on-2xl',
    mlXsOn_2xl: 'pf-u-ml-xs-on-2xl',
    mlSmOn_2xl: 'pf-u-ml-sm-on-2xl',
    mlMdOn_2xl: 'pf-u-ml-md-on-2xl',
    mlLgOn_2xl: 'pf-u-ml-lg-on-2xl',
    mlXlOn_2xl: 'pf-u-ml-xl-on-2xl',
    ml_2xlOn_2xl: 'pf-u-ml-2xl-on-2xl',
    ml_3xlOn_2xl: 'pf-u-ml-3xl-on-2xl',
    mxAutoOn_2xl: 'pf-u-mx-auto-on-2xl',
    mx_0On_2xl: 'pf-u-mx-0-on-2xl',
    mxXsOn_2xl: 'pf-u-mx-xs-on-2xl',
    mxSmOn_2xl: 'pf-u-mx-sm-on-2xl',
    mxMdOn_2xl: 'pf-u-mx-md-on-2xl',
    mxLgOn_2xl: 'pf-u-mx-lg-on-2xl',
    mxXlOn_2xl: 'pf-u-mx-xl-on-2xl',
    mx_2xlOn_2xl: 'pf-u-mx-2xl-on-2xl',
    mx_3xlOn_2xl: 'pf-u-mx-3xl-on-2xl',
    myAutoOn_2xl: 'pf-u-my-auto-on-2xl',
    my_0On_2xl: 'pf-u-my-0-on-2xl',
    myXsOn_2xl: 'pf-u-my-xs-on-2xl',
    mySmOn_2xl: 'pf-u-my-sm-on-2xl',
    myMdOn_2xl: 'pf-u-my-md-on-2xl',
    myLgOn_2xl: 'pf-u-my-lg-on-2xl',
    myXlOn_2xl: 'pf-u-my-xl-on-2xl',
    my_2xlOn_2xl: 'pf-u-my-2xl-on-2xl',
    my_3xlOn_2xl: 'pf-u-my-3xl-on-2xl',
    pAutoOn_2xl: 'pf-u-p-auto-on-2xl',
    p_0On_2xl: 'pf-u-p-0-on-2xl',
    pXsOn_2xl: 'pf-u-p-xs-on-2xl',
    pSmOn_2xl: 'pf-u-p-sm-on-2xl',
    pMdOn_2xl: 'pf-u-p-md-on-2xl',
    pLgOn_2xl: 'pf-u-p-lg-on-2xl',
    pXlOn_2xl: 'pf-u-p-xl-on-2xl',
    p_2xlOn_2xl: 'pf-u-p-2xl-on-2xl',
    p_3xlOn_2xl: 'pf-u-p-3xl-on-2xl',
    ptAutoOn_2xl: 'pf-u-pt-auto-on-2xl',
    pt_0On_2xl: 'pf-u-pt-0-on-2xl',
    ptXsOn_2xl: 'pf-u-pt-xs-on-2xl',
    ptSmOn_2xl: 'pf-u-pt-sm-on-2xl',
    ptMdOn_2xl: 'pf-u-pt-md-on-2xl',
    ptLgOn_2xl: 'pf-u-pt-lg-on-2xl',
    ptXlOn_2xl: 'pf-u-pt-xl-on-2xl',
    pt_2xlOn_2xl: 'pf-u-pt-2xl-on-2xl',
    pt_3xlOn_2xl: 'pf-u-pt-3xl-on-2xl',
    prAutoOn_2xl: 'pf-u-pr-auto-on-2xl',
    pr_0On_2xl: 'pf-u-pr-0-on-2xl',
    prXsOn_2xl: 'pf-u-pr-xs-on-2xl',
    prSmOn_2xl: 'pf-u-pr-sm-on-2xl',
    prMdOn_2xl: 'pf-u-pr-md-on-2xl',
    prLgOn_2xl: 'pf-u-pr-lg-on-2xl',
    prXlOn_2xl: 'pf-u-pr-xl-on-2xl',
    pr_2xlOn_2xl: 'pf-u-pr-2xl-on-2xl',
    pr_3xlOn_2xl: 'pf-u-pr-3xl-on-2xl',
    pbAutoOn_2xl: 'pf-u-pb-auto-on-2xl',
    pb_0On_2xl: 'pf-u-pb-0-on-2xl',
    pbXsOn_2xl: 'pf-u-pb-xs-on-2xl',
    pbSmOn_2xl: 'pf-u-pb-sm-on-2xl',
    pbMdOn_2xl: 'pf-u-pb-md-on-2xl',
    pbLgOn_2xl: 'pf-u-pb-lg-on-2xl',
    pbXlOn_2xl: 'pf-u-pb-xl-on-2xl',
    pb_2xlOn_2xl: 'pf-u-pb-2xl-on-2xl',
    pb_3xlOn_2xl: 'pf-u-pb-3xl-on-2xl',
    plAutoOn_2xl: 'pf-u-pl-auto-on-2xl',
    pl_0On_2xl: 'pf-u-pl-0-on-2xl',
    plXsOn_2xl: 'pf-u-pl-xs-on-2xl',
    plSmOn_2xl: 'pf-u-pl-sm-on-2xl',
    plMdOn_2xl: 'pf-u-pl-md-on-2xl',
    plLgOn_2xl: 'pf-u-pl-lg-on-2xl',
    plXlOn_2xl: 'pf-u-pl-xl-on-2xl',
    pl_2xlOn_2xl: 'pf-u-pl-2xl-on-2xl',
    pl_3xlOn_2xl: 'pf-u-pl-3xl-on-2xl',
    pxAutoOn_2xl: 'pf-u-px-auto-on-2xl',
    px_0On_2xl: 'pf-u-px-0-on-2xl',
    pxXsOn_2xl: 'pf-u-px-xs-on-2xl',
    pxSmOn_2xl: 'pf-u-px-sm-on-2xl',
    pxMdOn_2xl: 'pf-u-px-md-on-2xl',
    pxLgOn_2xl: 'pf-u-px-lg-on-2xl',
    pxXlOn_2xl: 'pf-u-px-xl-on-2xl',
    px_2xlOn_2xl: 'pf-u-px-2xl-on-2xl',
    px_3xlOn_2xl: 'pf-u-px-3xl-on-2xl',
    pyAutoOn_2xl: 'pf-u-py-auto-on-2xl',
    py_0On_2xl: 'pf-u-py-0-on-2xl',
    pyXsOn_2xl: 'pf-u-py-xs-on-2xl',
    pySmOn_2xl: 'pf-u-py-sm-on-2xl',
    pyMdOn_2xl: 'pf-u-py-md-on-2xl',
    pyLgOn_2xl: 'pf-u-py-lg-on-2xl',
    pyXlOn_2xl: 'pf-u-py-xl-on-2xl',
    py_2xlOn_2xl: 'pf-u-py-2xl-on-2xl',
    py_3xlOn_2xl: 'pf-u-py-3xl-on-2xl',
    modifiers: {}
};
