// @patternfly/react-styles/css/components/Select/select.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./select.css");
exports.default = {
    select: 'pf-c-select',
    selectToggle: 'pf-c-select__toggle',
    formControl: 'pf-c-form-control',
    selectToggleClear: 'pf-c-select__toggle-clear',
    selectToggleButton: 'pf-c-select__toggle-button',
    selectToggleArrow: 'pf-c-select__toggle-arrow',
    selectToggleText: 'pf-c-select__toggle-text',
    selectToggleWrapper: 'pf-c-select__toggle-wrapper',
    chip: 'pf-c-chip',
    button: 'pf-c-button',
    chipGroup: 'pf-c-chip-group',
    selectToggleTypeaheadForm: 'pf-c-select__toggle-typeahead-form',
    selectToggleTypeahead: 'pf-c-select__toggle-typeahead',
    selectToggleIcon: 'pf-c-select__toggle-icon',
    selectToggleBadge: 'pf-c-select__toggle-badge',
    selectMenu: 'pf-c-select__menu',
    selectMenuFieldset: 'pf-c-select__menu-fieldset',
    selectMenuItem: 'pf-c-select__menu-item',
    selectMenuItemIcon: 'pf-c-select__menu-item-icon',
    selectMenuItemMatch: 'pf-c-select__menu-item--match',
    selectSeparator: 'pf-c-select__separator',
    selectMenuGroup: 'pf-c-select__menu-group',
    selectMenuInput: 'pf-c-select__menu-input',
    selectMenuGroupTitle: 'pf-c-select__menu-group-title',
    modifiers: {
        'disabled': 'pf-m-disabled',
        'hover': 'pf-m-hover',
        'active': 'pf-m-active',
        'expanded': 'pf-m-expanded',
        'plain': 'pf-m-plain',
        'typeahead': 'pf-m-typeahead',
        'top': 'pf-m-top',
        'focus': 'pf-m-focus',
        'alignRight': 'pf-m-align-right',
        'selected': 'pf-m-selected'
    }
};
