// @patternfly/react-styles/css/components/NotificationDrawer/notification-drawer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./notification-drawer.css");
exports.default = {
    notificationDrawer: 'pf-c-notification-drawer',
    notificationDrawerHeader: 'pf-c-notification-drawer__header',
    notificationDrawerHeaderTitle: 'pf-c-notification-drawer__header-title',
    notificationDrawerHeaderStatus: 'pf-c-notification-drawer__header-status',
    notificationDrawerHeaderAction: 'pf-c-notification-drawer__header-action',
    notificationDrawerBody: 'pf-c-notification-drawer__body',
    notificationDrawerList: 'pf-c-notification-drawer__list',
    notificationDrawerListItem: 'pf-c-notification-drawer__list-item',
    notificationDrawerListItemHeader: 'pf-c-notification-drawer__list-item-header',
    notificationDrawerListItemHeaderIcon: 'pf-c-notification-drawer__list-item-header-icon',
    notificationDrawerListItemHeaderTitle: 'pf-c-notification-drawer__list-item-header-title',
    notificationDrawerListItemAction: 'pf-c-notification-drawer__list-item-action',
    notificationDrawerListItemDescription: 'pf-c-notification-drawer__list-item-description',
    notificationDrawerListItemTimestamp: 'pf-c-notification-drawer__list-item-timestamp',
    notificationDrawerGroupList: 'pf-c-notification-drawer__group-list',
    notificationDrawerGroup: 'pf-c-notification-drawer__group',
    notificationDrawerGroupToggle: 'pf-c-notification-drawer__group-toggle',
    notificationDrawerGroupToggleCount: 'pf-c-notification-drawer__group-toggle-count',
    notificationDrawerGroupToggleIcon: 'pf-c-notification-drawer__group-toggle-icon',
    modifiers: {
        'read': 'pf-m-read',
        'info': 'pf-m-info',
        'warning': 'pf-m-warning',
        'danger': 'pf-m-danger',
        'success': 'pf-m-success',
        'hoverable': 'pf-m-hoverable',
        'expanded': 'pf-m-expanded'
    }
};
