// @patternfly/react-styles/css/utilities/Flex/flex.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./flex.css");
exports.default = {
    flexDirectionColumn: 'pf-u-flex-direction-column',
    flexDirectionColumnReverse: 'pf-u-flex-direction-column-reverse',
    flexDirectionRow: 'pf-u-flex-direction-row',
    flexDirectionRowReverse: 'pf-u-flex-direction-row-reverse',
    flexDirectionColumnOnSm: 'pf-u-flex-direction-column-on-sm',
    flexDirectionColumnReverseOnSm: 'pf-u-flex-direction-column-reverse-on-sm',
    flexDirectionRowOnSm: 'pf-u-flex-direction-row-on-sm',
    flexDirectionRowReverseOnSm: 'pf-u-flex-direction-row-reverse-on-sm',
    flexDirectionColumnOnMd: 'pf-u-flex-direction-column-on-md',
    flexDirectionColumnReverseOnMd: 'pf-u-flex-direction-column-reverse-on-md',
    flexDirectionRowOnMd: 'pf-u-flex-direction-row-on-md',
    flexDirectionRowReverseOnMd: 'pf-u-flex-direction-row-reverse-on-md',
    flexDirectionColumnOnLg: 'pf-u-flex-direction-column-on-lg',
    flexDirectionColumnReverseOnLg: 'pf-u-flex-direction-column-reverse-on-lg',
    flexDirectionRowOnLg: 'pf-u-flex-direction-row-on-lg',
    flexDirectionRowReverseOnLg: 'pf-u-flex-direction-row-reverse-on-lg',
    flexDirectionColumnOnXl: 'pf-u-flex-direction-column-on-xl',
    flexDirectionColumnReverseOnXl: 'pf-u-flex-direction-column-reverse-on-xl',
    flexDirectionRowOnXl: 'pf-u-flex-direction-row-on-xl',
    flexDirectionRowReverseOnXl: 'pf-u-flex-direction-row-reverse-on-xl',
    flexDirectionColumnOn_2xl: 'pf-u-flex-direction-column-on-2xl',
    flexDirectionColumnReverseOn_2xl: 'pf-u-flex-direction-column-reverse-on-2xl',
    flexDirectionRowOn_2xl: 'pf-u-flex-direction-row-on-2xl',
    flexDirectionRowReverseOn_2xl: 'pf-u-flex-direction-row-reverse-on-2xl',
    flexWrap: 'pf-u-flex-wrap',
    flexNowrap: 'pf-u-flex-nowrap',
    flexWrapReverse: 'pf-u-flex-wrap-reverse',
    flexWrapOnSm: 'pf-u-flex-wrap-on-sm',
    flexNowrapOnSm: 'pf-u-flex-nowrap-on-sm',
    flexWrapReverseOnSm: 'pf-u-flex-wrap-reverse-on-sm',
    flexWrapOnMd: 'pf-u-flex-wrap-on-md',
    flexNowrapOnMd: 'pf-u-flex-nowrap-on-md',
    flexWrapReverseOnMd: 'pf-u-flex-wrap-reverse-on-md',
    flexWrapOnLg: 'pf-u-flex-wrap-on-lg',
    flexNowrapOnLg: 'pf-u-flex-nowrap-on-lg',
    flexWrapReverseOnLg: 'pf-u-flex-wrap-reverse-on-lg',
    flexWrapOnXl: 'pf-u-flex-wrap-on-xl',
    flexNowrapOnXl: 'pf-u-flex-nowrap-on-xl',
    flexWrapReverseOnXl: 'pf-u-flex-wrap-reverse-on-xl',
    flexWrapOn_2xl: 'pf-u-flex-wrap-on-2xl',
    flexNowrapOn_2xl: 'pf-u-flex-nowrap-on-2xl',
    flexWrapReverseOn_2xl: 'pf-u-flex-wrap-reverse-on-2xl',
    alignItemsFlexStart: 'pf-u-align-items-flex-start',
    alignItemsFlexEnd: 'pf-u-align-items-flex-end',
    alignItemsCenter: 'pf-u-align-items-center',
    alignItemsBaseline: 'pf-u-align-items-baseline',
    alignItemsStretch: 'pf-u-align-items-stretch',
    alignItemsFlexStartOnSm: 'pf-u-align-items-flex-start-on-sm',
    alignItemsFlexEndOnSm: 'pf-u-align-items-flex-end-on-sm',
    alignItemsCenterOnSm: 'pf-u-align-items-center-on-sm',
    alignItemsBaselineOnSm: 'pf-u-align-items-baseline-on-sm',
    alignItemsStretchOnSm: 'pf-u-align-items-stretch-on-sm',
    alignItemsFlexStartOnMd: 'pf-u-align-items-flex-start-on-md',
    alignItemsFlexEndOnMd: 'pf-u-align-items-flex-end-on-md',
    alignItemsCenterOnMd: 'pf-u-align-items-center-on-md',
    alignItemsBaselineOnMd: 'pf-u-align-items-baseline-on-md',
    alignItemsStretchOnMd: 'pf-u-align-items-stretch-on-md',
    alignItemsFlexStartOnLg: 'pf-u-align-items-flex-start-on-lg',
    alignItemsFlexEndOnLg: 'pf-u-align-items-flex-end-on-lg',
    alignItemsCenterOnLg: 'pf-u-align-items-center-on-lg',
    alignItemsBaselineOnLg: 'pf-u-align-items-baseline-on-lg',
    alignItemsStretchOnLg: 'pf-u-align-items-stretch-on-lg',
    alignItemsFlexStartOnXl: 'pf-u-align-items-flex-start-on-xl',
    alignItemsFlexEndOnXl: 'pf-u-align-items-flex-end-on-xl',
    alignItemsCenterOnXl: 'pf-u-align-items-center-on-xl',
    alignItemsBaselineOnXl: 'pf-u-align-items-baseline-on-xl',
    alignItemsStretchOnXl: 'pf-u-align-items-stretch-on-xl',
    alignItemsFlexStartOn_2xl: 'pf-u-align-items-flex-start-on-2xl',
    alignItemsFlexEndOn_2xl: 'pf-u-align-items-flex-end-on-2xl',
    alignItemsCenterOn_2xl: 'pf-u-align-items-center-on-2xl',
    alignItemsBaselineOn_2xl: 'pf-u-align-items-baseline-on-2xl',
    alignItemsStretchOn_2xl: 'pf-u-align-items-stretch-on-2xl',
    alignSelfFlexStart: 'pf-u-align-self-flex-start',
    alignSelfFlexEnd: 'pf-u-align-self-flex-end',
    alignSelfCenter: 'pf-u-align-self-center',
    alignSelfBaseline: 'pf-u-align-self-baseline',
    alignSelfStretch: 'pf-u-align-self-stretch',
    alignSelfFlexStartOnSm: 'pf-u-align-self-flex-start-on-sm',
    alignSelfFlexEndOnSm: 'pf-u-align-self-flex-end-on-sm',
    alignSelfCenterOnSm: 'pf-u-align-self-center-on-sm',
    alignSelfBaselineOnSm: 'pf-u-align-self-baseline-on-sm',
    alignSelfStretchOnSm: 'pf-u-align-self-stretch-on-sm',
    alignSelfFlexStartOnMd: 'pf-u-align-self-flex-start-on-md',
    alignSelfFlexEndOnMd: 'pf-u-align-self-flex-end-on-md',
    alignSelfCenterOnMd: 'pf-u-align-self-center-on-md',
    alignSelfBaselineOnMd: 'pf-u-align-self-baseline-on-md',
    alignSelfStretchOnMd: 'pf-u-align-self-stretch-on-md',
    alignSelfFlexStartOnLg: 'pf-u-align-self-flex-start-on-lg',
    alignSelfFlexEndOnLg: 'pf-u-align-self-flex-end-on-lg',
    alignSelfCenterOnLg: 'pf-u-align-self-center-on-lg',
    alignSelfBaselineOnLg: 'pf-u-align-self-baseline-on-lg',
    alignSelfStretchOnLg: 'pf-u-align-self-stretch-on-lg',
    alignSelfFlexStartOnXl: 'pf-u-align-self-flex-start-on-xl',
    alignSelfFlexEndOnXl: 'pf-u-align-self-flex-end-on-xl',
    alignSelfCenterOnXl: 'pf-u-align-self-center-on-xl',
    alignSelfBaselineOnXl: 'pf-u-align-self-baseline-on-xl',
    alignSelfStretchOnXl: 'pf-u-align-self-stretch-on-xl',
    alignSelfFlexStartOn_2xl: 'pf-u-align-self-flex-start-on-2xl',
    alignSelfFlexEndOn_2xl: 'pf-u-align-self-flex-end-on-2xl',
    alignSelfCenterOn_2xl: 'pf-u-align-self-center-on-2xl',
    alignSelfBaselineOn_2xl: 'pf-u-align-self-baseline-on-2xl',
    alignSelfStretchOn_2xl: 'pf-u-align-self-stretch-on-2xl',
    alignContentFlexStart: 'pf-u-align-content-flex-start',
    alignContentFlexEnd: 'pf-u-align-content-flex-end',
    alignContentCenter: 'pf-u-align-content-center',
    alignContentSpaceBetween: 'pf-u-align-content-space-between',
    alignContentSpaceAround: 'pf-u-align-content-space-around',
    alignContentStretch: 'pf-u-align-content-stretch',
    alignContentFlexStartOnSm: 'pf-u-align-content-flex-start-on-sm',
    alignContentFlexEndOnSm: 'pf-u-align-content-flex-end-on-sm',
    alignContentCenterOnSm: 'pf-u-align-content-center-on-sm',
    alignContentSpaceBetweenOnSm: 'pf-u-align-content-space-between-on-sm',
    alignContentSpaceAroundOnSm: 'pf-u-align-content-space-around-on-sm',
    alignContentStretchOnSm: 'pf-u-align-content-stretch-on-sm',
    alignContentFlexStartOnMd: 'pf-u-align-content-flex-start-on-md',
    alignContentFlexEndOnMd: 'pf-u-align-content-flex-end-on-md',
    alignContentCenterOnMd: 'pf-u-align-content-center-on-md',
    alignContentSpaceBetweenOnMd: 'pf-u-align-content-space-between-on-md',
    alignContentSpaceAroundOnMd: 'pf-u-align-content-space-around-on-md',
    alignContentStretchOnMd: 'pf-u-align-content-stretch-on-md',
    alignContentFlexStartOnLg: 'pf-u-align-content-flex-start-on-lg',
    alignContentFlexEndOnLg: 'pf-u-align-content-flex-end-on-lg',
    alignContentCenterOnLg: 'pf-u-align-content-center-on-lg',
    alignContentSpaceBetweenOnLg: 'pf-u-align-content-space-between-on-lg',
    alignContentSpaceAroundOnLg: 'pf-u-align-content-space-around-on-lg',
    alignContentStretchOnLg: 'pf-u-align-content-stretch-on-lg',
    alignContentFlexStartOnXl: 'pf-u-align-content-flex-start-on-xl',
    alignContentFlexEndOnXl: 'pf-u-align-content-flex-end-on-xl',
    alignContentCenterOnXl: 'pf-u-align-content-center-on-xl',
    alignContentSpaceBetweenOnXl: 'pf-u-align-content-space-between-on-xl',
    alignContentSpaceAroundOnXl: 'pf-u-align-content-space-around-on-xl',
    alignContentStretchOnXl: 'pf-u-align-content-stretch-on-xl',
    alignContentFlexStartOn_2xl: 'pf-u-align-content-flex-start-on-2xl',
    alignContentFlexEndOn_2xl: 'pf-u-align-content-flex-end-on-2xl',
    alignContentCenterOn_2xl: 'pf-u-align-content-center-on-2xl',
    alignContentSpaceBetweenOn_2xl: 'pf-u-align-content-space-between-on-2xl',
    alignContentSpaceAroundOn_2xl: 'pf-u-align-content-space-around-on-2xl',
    alignContentStretchOn_2xl: 'pf-u-align-content-stretch-on-2xl',
    justifyContentFlexStart: 'pf-u-justify-content-flex-start',
    justifyContentFlexEnd: 'pf-u-justify-content-flex-end',
    justifyContentCenter: 'pf-u-justify-content-center',
    justifyContentSpaceBetween: 'pf-u-justify-content-space-between',
    justifyContentSpaceAround: 'pf-u-justify-content-space-around',
    justifyContentStretch: 'pf-u-justify-content-stretch',
    justifyContentFlexStartOnSm: 'pf-u-justify-content-flex-start-on-sm',
    justifyContentFlexEndOnSm: 'pf-u-justify-content-flex-end-on-sm',
    justifyContentCenterOnSm: 'pf-u-justify-content-center-on-sm',
    justifyContentSpaceBetweenOnSm: 'pf-u-justify-content-space-between-on-sm',
    justifyContentSpaceAroundOnSm: 'pf-u-justify-content-space-around-on-sm',
    justifyContentStretchOnSm: 'pf-u-justify-content-stretch-on-sm',
    justifyContentFlexStartOnMd: 'pf-u-justify-content-flex-start-on-md',
    justifyContentFlexEndOnMd: 'pf-u-justify-content-flex-end-on-md',
    justifyContentCenterOnMd: 'pf-u-justify-content-center-on-md',
    justifyContentSpaceBetweenOnMd: 'pf-u-justify-content-space-between-on-md',
    justifyContentSpaceAroundOnMd: 'pf-u-justify-content-space-around-on-md',
    justifyContentStretchOnMd: 'pf-u-justify-content-stretch-on-md',
    justifyContentFlexStartOnLg: 'pf-u-justify-content-flex-start-on-lg',
    justifyContentFlexEndOnLg: 'pf-u-justify-content-flex-end-on-lg',
    justifyContentCenterOnLg: 'pf-u-justify-content-center-on-lg',
    justifyContentSpaceBetweenOnLg: 'pf-u-justify-content-space-between-on-lg',
    justifyContentSpaceAroundOnLg: 'pf-u-justify-content-space-around-on-lg',
    justifyContentStretchOnLg: 'pf-u-justify-content-stretch-on-lg',
    justifyContentFlexStartOnXl: 'pf-u-justify-content-flex-start-on-xl',
    justifyContentFlexEndOnXl: 'pf-u-justify-content-flex-end-on-xl',
    justifyContentCenterOnXl: 'pf-u-justify-content-center-on-xl',
    justifyContentSpaceBetweenOnXl: 'pf-u-justify-content-space-between-on-xl',
    justifyContentSpaceAroundOnXl: 'pf-u-justify-content-space-around-on-xl',
    justifyContentStretchOnXl: 'pf-u-justify-content-stretch-on-xl',
    justifyContentFlexStartOn_2xl: 'pf-u-justify-content-flex-start-on-2xl',
    justifyContentFlexEndOn_2xl: 'pf-u-justify-content-flex-end-on-2xl',
    justifyContentCenterOn_2xl: 'pf-u-justify-content-center-on-2xl',
    justifyContentSpaceBetweenOn_2xl: 'pf-u-justify-content-space-between-on-2xl',
    justifyContentSpaceAroundOn_2xl: 'pf-u-justify-content-space-around-on-2xl',
    justifyContentStretchOn_2xl: 'pf-u-justify-content-stretch-on-2xl',
    flexShrink_1: 'pf-u-flex-shrink-1',
    flexGrow_1: 'pf-u-flex-grow-1',
    flexShrink_0: 'pf-u-flex-shrink-0',
    flexGrow_0: 'pf-u-flex-grow-0',
    flexShrink_1OnSm: 'pf-u-flex-shrink-1-on-sm',
    flexGrow_1OnSm: 'pf-u-flex-grow-1-on-sm',
    flexShrink_0OnSm: 'pf-u-flex-shrink-0-on-sm',
    flexGrow_0OnSm: 'pf-u-flex-grow-0-on-sm',
    flexShrink_1OnMd: 'pf-u-flex-shrink-1-on-md',
    flexGrow_1OnMd: 'pf-u-flex-grow-1-on-md',
    flexShrink_0OnMd: 'pf-u-flex-shrink-0-on-md',
    flexGrow_0OnMd: 'pf-u-flex-grow-0-on-md',
    flexShrink_1OnLg: 'pf-u-flex-shrink-1-on-lg',
    flexGrow_1OnLg: 'pf-u-flex-grow-1-on-lg',
    flexShrink_0OnLg: 'pf-u-flex-shrink-0-on-lg',
    flexGrow_0OnLg: 'pf-u-flex-grow-0-on-lg',
    flexShrink_1OnXl: 'pf-u-flex-shrink-1-on-xl',
    flexGrow_1OnXl: 'pf-u-flex-grow-1-on-xl',
    flexShrink_0OnXl: 'pf-u-flex-shrink-0-on-xl',
    flexGrow_0OnXl: 'pf-u-flex-grow-0-on-xl',
    flexShrink_1On_2xl: 'pf-u-flex-shrink-1-on-2xl',
    flexGrow_1On_2xl: 'pf-u-flex-grow-1-on-2xl',
    flexShrink_0On_2xl: 'pf-u-flex-shrink-0-on-2xl',
    flexGrow_0On_2xl: 'pf-u-flex-grow-0-on-2xl',
    flexBasis_0: 'pf-u-flex-basis-0',
    flexBasisAuto: 'pf-u-flex-basis-auto',
    flexBasis_0OnSm: 'pf-u-flex-basis-0-on-sm',
    flexBasisAutoOnSm: 'pf-u-flex-basis-auto-on-sm',
    flexBasis_0OnMd: 'pf-u-flex-basis-0-on-md',
    flexBasisAutoOnMd: 'pf-u-flex-basis-auto-on-md',
    flexBasis_0OnLg: 'pf-u-flex-basis-0-on-lg',
    flexBasisAutoOnLg: 'pf-u-flex-basis-auto-on-lg',
    flexBasis_0OnXl: 'pf-u-flex-basis-0-on-xl',
    flexBasisAutoOnXl: 'pf-u-flex-basis-auto-on-xl',
    flexBasis_0On_2xl: 'pf-u-flex-basis-0-on-2xl',
    flexBasisAutoOn_2xl: 'pf-u-flex-basis-auto-on-2xl',
    flexNone: 'pf-u-flex-none',
    flex_1: 'pf-u-flex-1',
    flexNoneOnSm: 'pf-u-flex-none-on-sm',
    flex_1OnSm: 'pf-u-flex-1-on-sm',
    flexNoneOnMd: 'pf-u-flex-none-on-md',
    flex_1OnMd: 'pf-u-flex-1-on-md',
    flexNoneOnLg: 'pf-u-flex-none-on-lg',
    flex_1OnLg: 'pf-u-flex-1-on-lg',
    flexNoneOnXl: 'pf-u-flex-none-on-xl',
    flex_1OnXl: 'pf-u-flex-1-on-xl',
    flexNoneOn_2xl: 'pf-u-flex-none-on-2xl',
    flex_1On_2xl: 'pf-u-flex-1-on-2xl',
    flexFill: 'pf-u-flex-fill',
    flexFillOnSm: 'pf-u-flex-fill-on-sm',
    flexFillOnMd: 'pf-u-flex-fill-on-md',
    flexFillOnLg: 'pf-u-flex-fill-on-lg',
    flexFillOnXl: 'pf-u-flex-fill-on-xl',
    flexFillOn_2xl: 'pf-u-flex-fill-on-2xl',
    modifiers: {}
};
