// @patternfly/react-styles/css/utilities/Display/display.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./display.css");
exports.default = {
    displayNone: 'pf-u-display-none',
    displayInlineBlock: 'pf-u-display-inline-block',
    displayBlock: 'pf-u-display-block',
    displayInline: 'pf-u-display-inline',
    displayTable: 'pf-u-display-table',
    displayTableCell: 'pf-u-display-table-cell',
    displayTableRow: 'pf-u-display-table-row',
    displayFlex: 'pf-u-display-flex',
    displayInlineFlex: 'pf-u-display-inline-flex',
    displayNoneOnSm: 'pf-u-display-none-on-sm',
    displayInlineBlockOnSm: 'pf-u-display-inline-block-on-sm',
    displayBlockOnSm: 'pf-u-display-block-on-sm',
    displayInlineOnSm: 'pf-u-display-inline-on-sm',
    displayTableOnSm: 'pf-u-display-table-on-sm',
    displayTableCellOnSm: 'pf-u-display-table-cell-on-sm',
    displayTableRowOnSm: 'pf-u-display-table-row-on-sm',
    displayFlexOnSm: 'pf-u-display-flex-on-sm',
    displayInlineFlexOnSm: 'pf-u-display-inline-flex-on-sm',
    displayNoneOnMd: 'pf-u-display-none-on-md',
    displayInlineBlockOnMd: 'pf-u-display-inline-block-on-md',
    displayBlockOnMd: 'pf-u-display-block-on-md',
    displayInlineOnMd: 'pf-u-display-inline-on-md',
    displayTableOnMd: 'pf-u-display-table-on-md',
    displayTableCellOnMd: 'pf-u-display-table-cell-on-md',
    displayTableRowOnMd: 'pf-u-display-table-row-on-md',
    displayFlexOnMd: 'pf-u-display-flex-on-md',
    displayInlineFlexOnMd: 'pf-u-display-inline-flex-on-md',
    displayNoneOnLg: 'pf-u-display-none-on-lg',
    displayInlineBlockOnLg: 'pf-u-display-inline-block-on-lg',
    displayBlockOnLg: 'pf-u-display-block-on-lg',
    displayInlineOnLg: 'pf-u-display-inline-on-lg',
    displayTableOnLg: 'pf-u-display-table-on-lg',
    displayTableCellOnLg: 'pf-u-display-table-cell-on-lg',
    displayTableRowOnLg: 'pf-u-display-table-row-on-lg',
    displayFlexOnLg: 'pf-u-display-flex-on-lg',
    displayInlineFlexOnLg: 'pf-u-display-inline-flex-on-lg',
    displayNoneOnXl: 'pf-u-display-none-on-xl',
    displayInlineBlockOnXl: 'pf-u-display-inline-block-on-xl',
    displayBlockOnXl: 'pf-u-display-block-on-xl',
    displayInlineOnXl: 'pf-u-display-inline-on-xl',
    displayTableOnXl: 'pf-u-display-table-on-xl',
    displayTableCellOnXl: 'pf-u-display-table-cell-on-xl',
    displayTableRowOnXl: 'pf-u-display-table-row-on-xl',
    displayFlexOnXl: 'pf-u-display-flex-on-xl',
    displayInlineFlexOnXl: 'pf-u-display-inline-flex-on-xl',
    displayNoneOn_2xl: 'pf-u-display-none-on-2xl',
    displayInlineBlockOn_2xl: 'pf-u-display-inline-block-on-2xl',
    displayBlockOn_2xl: 'pf-u-display-block-on-2xl',
    displayInlineOn_2xl: 'pf-u-display-inline-on-2xl',
    displayTableOn_2xl: 'pf-u-display-table-on-2xl',
    displayTableCellOn_2xl: 'pf-u-display-table-cell-on-2xl',
    displayTableRowOn_2xl: 'pf-u-display-table-row-on-2xl',
    displayFlexOn_2xl: 'pf-u-display-flex-on-2xl',
    displayInlineFlexOn_2xl: 'pf-u-display-inline-flex-on-2xl',
    modifiers: {}
};
