// @patternfly/react-styles/css/components/Dropdown/dropdown.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
require("./dropdown.css");
exports.default = {
    dropdown: 'pf-c-dropdown',
    divider: 'pf-c-divider',
    dropdownToggle: 'pf-c-dropdown__toggle',
    dropdownToggleButton: 'pf-c-dropdown__toggle-button',
    dropdownToggleCheck: 'pf-c-dropdown__toggle-check',
    dropdownToggleText: 'pf-c-dropdown__toggle-text',
    dropdownToggleIcon: 'pf-c-dropdown__toggle-icon',
    dropdownMenu: 'pf-c-dropdown__menu',
    dropdownMenuItem: 'pf-c-dropdown__menu-item',
    dropdownMenuItemIcon: 'pf-c-dropdown__menu-item-icon',
    dropdownSeparator: 'pf-c-dropdown__separator',
    dropdownGroup: 'pf-c-dropdown__group',
    dropdownGroupTitle: 'pf-c-dropdown__group-title',
    modifiers: {
        'disabled': 'pf-m-disabled',
        'plain': 'pf-m-plain',
        'splitButton': 'pf-m-split-button',
        'action': 'pf-m-action',
        'hover': 'pf-m-hover',
        'active': 'pf-m-active',
        'focus': 'pf-m-focus',
        'expanded': 'pf-m-expanded',
        'primary': 'pf-m-primary',
        'top': 'pf-m-top',
        'alignRight': 'pf-m-align-right',
        'icon': 'pf-m-icon'
    }
};
